{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57d97be5",
   "metadata": {},
   "source": [
    "\n",
    "## Vignette for extracting Long Ids\n",
    "Reference Guided Isoform Discovery (Bulk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyranges as pr\n",
    "import gffutils\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_colwidth', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2",
   "metadata": {},
   "source": [
    "### Matching gene and and transcript ids from Gffcompare outs with Isoquant outs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "!which conda"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "<h3>looking at the tracking file from gffcompare outs for matching ids </h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "wd=\"../test_data/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames=[\"query_transfrag_id\", \"query_locus_id\", \"reference_gene_id\",\"class_code\",\"qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len\"]\n",
    "tracking_df = pd.read_csv(wd+'noRef_isoformSwitchR/gffcmp.tracking', sep='\\t',names=colnames, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(59734, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_transfrag_id</th>\n",
       "      <th>query_locus_id</th>\n",
       "      <th>reference_gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCONS_00000001</td>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>q1:MSTRG.49|MSTRG.49.1|4|0.000000|0.000000|0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCONS_00000002</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>q1:MSTRG.62|MSTRG.62.1|8|0.000000|0.000000|0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCONS_00000003</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>q1:MSTRG.62|MSTRG.62.2|8|0.000000|0.000000|0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCONS_00000004</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>q1:MSTRG.62|MSTRG.62.3|8|0.000000|0.000000|0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCONS_00000005</td>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>q1:MSTRG.52|MSTRG.52.1|4|0.000000|0.000000|0.0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_transfrag_id query_locus_id reference_gene_id class_code  \\\n",
       "0     TCONS_00000001    XLOC_000001                 -          u   \n",
       "1     TCONS_00000002    XLOC_000002                 -          u   \n",
       "2     TCONS_00000003    XLOC_000002                 -          u   \n",
       "3     TCONS_00000004    XLOC_000002                 -          u   \n",
       "4     TCONS_00000005    XLOC_000003                 -          u   \n",
       "\n",
       "  qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len  \n",
       "0  q1:MSTRG.49|MSTRG.49.1|4|0.000000|0.000000|0.0...   \n",
       "1  q1:MSTRG.62|MSTRG.62.1|8|0.000000|0.000000|0.0...   \n",
       "2  q1:MSTRG.62|MSTRG.62.2|8|0.000000|0.000000|0.0...   \n",
       "3  q1:MSTRG.62|MSTRG.62.3|8|0.000000|0.000000|0.0...   \n",
       "4  q1:MSTRG.52|MSTRG.52.1|4|0.000000|0.000000|0.0...   "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(tracking_df.shape)\n",
    "tracking_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#col0 gene_ids qJ:gene_id\n",
    "tracking_df[\"gene_id\"]=tracking_df['qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len'].apply(lambda x: x.split('|')[0].split(':')[1])\n",
    "#col1 transcript_id\n",
    "tracking_df[\"transcript_id\"]=tracking_df['qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len'].apply(lambda x: x.split('|')[1])\n",
    "#col2 num_exons\n",
    "tracking_df[\"num_exons\"]=tracking_df['qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len'].apply(lambda x: x.split('|')[2])\n",
    "#col3 FPKM\n",
    "tracking_df['qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len'].apply(lambda x: x.split('|')[3]).unique()\n",
    "#col4 TPM\n",
    "tracking_df['qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len'].apply(lambda x: x.split('|')[4]).unique()\n",
    "#col5 cov\n",
    "tracking_df['qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len'].apply(lambda x: x.split('|')[5]).unique()\n",
    "#col6 len\n",
    "tracking_df[\"len\"]=tracking_df['qJ:gene_id|transcript_id|num_exons|FPKM|TPM|cov|len'].apply(lambda x: x.split('|')[6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "tracking_df=tracking_df[['query_transfrag_id','query_locus_id','reference_gene_id','class_code','gene_id','transcript_id','num_exons','len']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "57791d42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(59734,)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tracking_df.transcript_id.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_transfrag_id</th>\n",
       "      <th>query_locus_id</th>\n",
       "      <th>reference_gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_exons</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCONS_00000001</td>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>4</td>\n",
       "      <td>1035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCONS_00000002</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.62</td>\n",
       "      <td>MSTRG.62.1</td>\n",
       "      <td>8</td>\n",
       "      <td>2579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCONS_00000003</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.62</td>\n",
       "      <td>MSTRG.62.2</td>\n",
       "      <td>8</td>\n",
       "      <td>2541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCONS_00000004</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.62</td>\n",
       "      <td>MSTRG.62.3</td>\n",
       "      <td>8</td>\n",
       "      <td>2512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCONS_00000005</td>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.52</td>\n",
       "      <td>MSTRG.52.1</td>\n",
       "      <td>4</td>\n",
       "      <td>4605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59729</th>\n",
       "      <td>TCONS_00059730</td>\n",
       "      <td>XLOC_027554</td>\n",
       "      <td>ENSMUSG00000068457.15|ENSMUST00000143958.8</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.29660</td>\n",
       "      <td>MSTRG.29660.1</td>\n",
       "      <td>6</td>\n",
       "      <td>1310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59730</th>\n",
       "      <td>TCONS_00059731</td>\n",
       "      <td>XLOC_027557</td>\n",
       "      <td>ENSMUSG00000069045.12|ENSMUST00000091190.12</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.29662</td>\n",
       "      <td>MSTRG.29662.1</td>\n",
       "      <td>17</td>\n",
       "      <td>4694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59731</th>\n",
       "      <td>TCONS_00059732</td>\n",
       "      <td>XLOC_027557</td>\n",
       "      <td>ENSMUSG00000069045.12|ENSMUST00000091190.12</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.29662</td>\n",
       "      <td>MSTRG.29662.2</td>\n",
       "      <td>17</td>\n",
       "      <td>3851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59732</th>\n",
       "      <td>TCONS_00059733</td>\n",
       "      <td>XLOC_027557</td>\n",
       "      <td>ENSMUSG00000069045.12|ENSMUST00000188182.7</td>\n",
       "      <td>k</td>\n",
       "      <td>MSTRG.29662</td>\n",
       "      <td>MSTRG.29662.3</td>\n",
       "      <td>16</td>\n",
       "      <td>3914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59733</th>\n",
       "      <td>TCONS_00059734</td>\n",
       "      <td>XLOC_027558</td>\n",
       "      <td>ENSMUSG00000095366.3|ENSMUST00000177893.2</td>\n",
       "      <td>k</td>\n",
       "      <td>MSTRG.29666</td>\n",
       "      <td>MSTRG.29666.1</td>\n",
       "      <td>4</td>\n",
       "      <td>1469</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59734 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      query_transfrag_id query_locus_id  \\\n",
       "0         TCONS_00000001    XLOC_000001   \n",
       "1         TCONS_00000002    XLOC_000002   \n",
       "2         TCONS_00000003    XLOC_000002   \n",
       "3         TCONS_00000004    XLOC_000002   \n",
       "4         TCONS_00000005    XLOC_000003   \n",
       "...                  ...            ...   \n",
       "59729     TCONS_00059730    XLOC_027554   \n",
       "59730     TCONS_00059731    XLOC_027557   \n",
       "59731     TCONS_00059732    XLOC_027557   \n",
       "59732     TCONS_00059733    XLOC_027557   \n",
       "59733     TCONS_00059734    XLOC_027558   \n",
       "\n",
       "                                 reference_gene_id class_code      gene_id  \\\n",
       "0                                                -          u     MSTRG.49   \n",
       "1                                                -          u     MSTRG.62   \n",
       "2                                                -          u     MSTRG.62   \n",
       "3                                                -          u     MSTRG.62   \n",
       "4                                                -          u     MSTRG.52   \n",
       "...                                            ...        ...          ...   \n",
       "59729   ENSMUSG00000068457.15|ENSMUST00000143958.8          j  MSTRG.29660   \n",
       "59730  ENSMUSG00000069045.12|ENSMUST00000091190.12          =  MSTRG.29662   \n",
       "59731  ENSMUSG00000069045.12|ENSMUST00000091190.12          j  MSTRG.29662   \n",
       "59732   ENSMUSG00000069045.12|ENSMUST00000188182.7          k  MSTRG.29662   \n",
       "59733    ENSMUSG00000095366.3|ENSMUST00000177893.2          k  MSTRG.29666   \n",
       "\n",
       "       transcript_id num_exons   len  \n",
       "0         MSTRG.49.1         4  1035  \n",
       "1         MSTRG.62.1         8  2579  \n",
       "2         MSTRG.62.2         8  2541  \n",
       "3         MSTRG.62.3         8  2512  \n",
       "4         MSTRG.52.1         4  4605  \n",
       "...              ...       ...   ...  \n",
       "59729  MSTRG.29660.1         6  1310  \n",
       "59730  MSTRG.29662.1        17  4694  \n",
       "59731  MSTRG.29662.2        17  3851  \n",
       "59732  MSTRG.29662.3        16  3914  \n",
       "59733  MSTRG.29666.1         4  1469  \n",
       "\n",
       "[59734 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tracking_df\n",
    "#check - gffcompare matching at gene vs transcript level"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12",
   "metadata": {},
   "source": [
    "### Looking at the distribution of class codes (similar to SQANTI)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "14",
   "metadata": {},
   "source": [
    "<div class=\"xtable\"><table frame=\"hsides\" rules=\"groups\" content-type=\"article-table\" class=\"rendered small default_table\"><thead><tr><th align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">Code</th><th align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">Relationship to reference transcript</th></tr></thead><tbody><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">=</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">complete, exact intron chain match</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">c</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">contained in reference transcript (intron compatible)</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">k</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">contains reference transcript (reverse containment)</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">m</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">retained intron(s) compared to reference, full intron chain match\n",
    "<br>everywhere else</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">n</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">completely overlaps intron from reference transcript, partial or no intron\n",
    "<br>chain match everywhere else</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">j</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">multi-exon with at least one junction match</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">e</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">single exon that partially covers an intron from reference</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">o</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">other same strand overlap with reference exons</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">s</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">intron match on the opposite strand (likely a mapping error)</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">x</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">exonic overlap on the opposite strand</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">i</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">fully contained within a reference intron</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">y</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">contains a reference within its intron(s)</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">p</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">possible polymerase run-on (close to reference but no overlap)</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">r</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">repeat (at least 50% bases are soft-masked)</td></tr><tr><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">u</td><td align=\"left\" valign=\"top\" rowspan=\"1\" colspan=\"1\">none of the above (unknown, intergenic)</td></tr></tbody></table></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Text(0.5, 1.0, 'distribution of class codes - Gffcompare \\n Ref Free track SAryal data')]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAk0AAAHWCAYAAABqs71FAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAABdyUlEQVR4nO3deVhUZf8/8PfIMgwII8hebKEiCu4bUIKpIAlWWqYkaipuKbll4YqVmlvZ41JmKi6Y9pSVqQ8CmqaBe5gomZoIFrgiCCIg3L8/+nG+jgN4QGQGfb+u61w59/mcc99nZoB3Z1UIIQSIiIiIqEoNdD0AIiIiovqAoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSgaGJ6oXo6GgoFAqNNldXVwwdOrRa60lKSkJ0dDRu3bpVreUe7Gvfvn1QKBT49ttvq7Weqty5cwfR0dHYt2+f1ryYmBgoFAqkp6fXWn+Pw4wZM+Ds7AxDQ0M0atTokdZVX7b5cajo+66PLl68iMjISHh6esLMzAwmJiZwdXXFoEGD8PPPP+PBB05s3boVLVu2hEqlgkKhQEpKCgBg2bJlaNKkCYyNjaFQKKr980lUVwx1PQCimvr+++9hYWFRrWWSkpIwZ84cDB06tFp/1GvSV3XduXMHc+bMAQAEBARozOvduzeSk5Ph4ODwWMfwKH788UfMnTsX06dPR3BwMJRKpa6HRI/R9u3bERYWBmtra4wePRrt2rWDUqnE+fPn8e233+LFF19EYmIiunfvDgC4du0awsPD0atXL6xcuRJKpRLNmjVDSkoKIiMjMWLECAwZMgSGhoYwNzfX8dYRVYyhieqttm3bPvY+CgsLoVKp6qSvqtjY2MDGxkanY3iY1NRUAEBkZCRsbW11PBp6nC5cuICBAweiZcuWSExM1PgfCn9/fwwfPhz79u2DpaWl1P7nn3+ipKQEgwYNgr+/v9R++vRpAEBERAQ6depUdxuhhwoLC2FiYlIv9jI+rXh4jvTOzp070aZNGyiVSri5uWHx4sUV1j14yKysrAwfffQRPDw8oFKp0KhRI7Rq1QqfffYZgH8Pebz77rsAADc3NygUCigUCulwmKurK0JCQrBt2za0bdsWJiYm0p6fyg4F3r17F5MmTYK9vT1UKhX8/f3x22+/adQEBARo7TkCgKFDh8LV1RUAkJ6eLoWiOXPmSGMr77OyQ1Vr165F69atYWJiAisrK7z66qtIS0vT6qdhw4Y4f/48XnrpJTRs2BBOTk6YPHkyioqKKnxv71dWVoaFCxeiefPmUCqVsLW1xeDBg3H58mWpxtXVFTNmzAAA2NnZQaFQIDo6usr1Hj58GKGhoWjcuDFMTEzg7u6OCRMmVLlMQkICXn75ZTz77LMwMTFBkyZNMGrUKFy/fl2j7tq1axg5ciScnJygVCphY2MDPz8/JCYmSjW//fYbQkJCYGtrC6VSCUdHR/Tu3VtjuyoTFxeH7t27Q61Ww9TUFJ6enpg/f75Gzfbt2+Hj4wNTU1OYm5ujZ8+eSE5O1lqX3O+7EAIrV65EmzZtoFKpYGlpiddeew1//fWXRt2jbJdcn3zyCe7cuYOVK1dWugc2ICAArVu3BvDvd/D5558HALzxxhtQKBTSz8WgQYMAAJ07d9b4zgO19z6XH+78/fff8frrr0OtVsPKygqTJk3CvXv3cPbsWfTq1Qvm5uZwdXXFwoULNZYvPxy/adOmh/68Hzt2DAMGDICrqytUKhVcXV0xcOBAXLp0SaOu/Gc6Pj4ew4YNg42NDUxNTaWfya1bt8LHxwdmZmZo2LAhgoKCtPqiusc9TaRX9uzZg5dffhk+Pj7YsmULSktLsXDhQly5cuWhyy5cuBDR0dGYMWMGunbtipKSEvzxxx/S+REjRozAzZs3sWzZMmzbtk061NWiRQtpHSdOnEBaWhpmzJgBNzc3mJmZVdnntGnT0K5dO3z11VfIzc1FdHQ0AgIC8Ntvv+G5556Tvd0ODg6Ii4tDr169MHz4cIwYMQIAqty7NH/+fEybNg0DBw7E/PnzcePGDURHR8PHxwdHjx5F06ZNpdqSkhL06dMHw4cPx+TJk/HLL7/gww8/hFqtxqxZs6oc25gxY/Dll19i3LhxCAkJQXp6OmbOnIl9+/bhxIkTsLa2xvfff48VK1ZgzZo1iIuLg1qtxrPPPlvpOnfv3o3Q0FB4enrik08+gbOzM9LT0xEfH1/lWC5cuAAfHx+MGDECarUa6enp+OSTT/D888/j1KlTMDIyAgCEh4fjxIkTmDt3Lpo1a4Zbt27hxIkTuHHjBgCgoKAAPXv2hJubG1asWAE7OztkZ2fj559/xu3bt6scw5o1axAREQF/f3988cUXsLW1xZ9//intaQOAzZs3480330RgYCC+/vprFBUVYeHChQgICMCePXukAFGd7/uoUaMQExODyMhILFiwADdv3sQHH3wAX19fnDx5EnZ2do+0XdWRkJAABwcHdOjQQVb9zJkz0alTJ7z99tuYN28eunXrJoWtr7/+Gh999BHWrVuH5s2bS9/52nyfy/Xv3x+DBg3CqFGjkJCQgIULF6KkpASJiYkYO3YspkyZgs2bN+O9995DkyZN0LdvX43l5fy8p6enw8PDAwMGDICVlRWysrLw+eefo2PHjjhz5gysra011jls2DD07t0bGzduREFBAYyMjDBv3jzMmDEDb731FmbMmIHi4mIsWrQIL7zwAo4cOaLxO4vqmCDSI507dxaOjo6isLBQasvLyxNWVlbiwa+ri4uLGDJkiPQ6JCREtGnTpsr1L1q0SAAQFy9e1Jrn4uIiDAwMxNmzZyucd39fP//8swAg2rVrJ8rKyqT29PR0YWRkJEaMGCG1+fv7C39/f611DhkyRLi4uEivr127JgCI2bNna9WuW7dOY9w5OTlCpVKJl156SaMuIyNDKJVKERYWptEPAPHNN99o1L700kvCw8NDq6/7paWlCQBi7NixGu2HDx8WAMS0adOkttmzZwsA4tq1a1WuUwgh3N3dhbu7u8bn/KAHt/lBZWVloqSkRFy6dEkAED/++KM0r2HDhmLChAmVrvvYsWMCgPjhhx8eOtb73b59W1hYWIjnn39e43O/X2lpqXB0dBTe3t6itLRUY1lbW1vh6+srtcn9vicnJwsAYsmSJRp9ZWZmCpVKJaZOnfpI21VdJiYmokuXLlrtpaWloqSkRJru3/7yn5n//ve/GsuUf85Hjx6V2mr7fS7/bj74/rVp00YAENu2bZPaSkpKhI2Njejbt6/W2OX8vD/o3r17Ij8/X5iZmYnPPvtMa7sHDx6sUZ+RkSEMDQ3F+PHjNdpv374t7O3tRf/+/Svtix4/Hp4jvVFQUICjR4+ib9++MDExkdrNzc0RGhr60OU7deqEkydPYuzYsdi9ezfy8vKqPYZWrVqhWbNmsuvDwsI0zj9wcXGBr68vfv7552r3XR3JyckoLCzUOmTo5OSEF198EXv27NFoVygUWu9hq1attA4ZPKh8Ox7sp1OnTvD09NTqR44///wTFy5cwPDhwzU+ZzmuXr2K0aNHw8nJCYaGhjAyMoKLiwsAaByW7NSpE2JiYvDRRx/h0KFDKCkp0VhPkyZNYGlpiffeew9ffPEFzpw5I6v/pKQk5OXlYezYsZWed3L27Fn8888/CA8PR4MG//crtmHDhujXrx8OHTqEO3fuVOv7vmPHDigUCgwaNAj37t2TJnt7e7Ru3Vo6xFzT7RJCaKz33r17spZ7UN++fWFkZCRNkZGRNVpPbb7P9wsJCdF47enpCYVCgeDgYKnN0NAQTZo0qfBnQ87Pe35+vrSnytDQEIaGhmjYsCEKCgq0Dp0DQL9+/TRe7969G/fu3cPgwYM1Pg8TExP4+/tXeHUt1R2GJtIbOTk5KCsrg729vda8itoeFBUVhcWLF+PQoUMIDg5G48aN0b17dxw7dkz2GKp7dVplYy0/DPS4lK+/ovE6Ojpq9W9qaqoVUJRKJe7evVur/chx7do1AKjy8F1FysrKEBgYiG3btmHq1KnYs2cPjhw5gkOHDgH49yTaclu3bsWQIUPw1VdfwcfHB1ZWVhg8eDCys7MBAGq1Gvv370ebNm0wbdo0tGzZEo6Ojpg9e7ZWwKru2B/2npWVlSEnJ6da3/crV65ACAE7OzuNUGJkZIRDhw5J53TVdLvWr1+vtd6qODs7VxgqlixZgqNHj+Lo0aNVLv8wtfk+38/KykrjtbGxcYU/G8bGxhX+bMj5eQ8LC8Py5csxYsQI7N69G0eOHMHRo0dhY2Oj8R0t9+D4yw/NduzYUesz2bp1q9b5e1S3eE4T6Q1LS0soFArpD9v9Kmp7kKGhISZNmoRJkybh1q1bSExMxLRp0xAUFITMzEyYmpo+dB3VvWqlsrE2btxYem1iYoLc3Fytukf55Ve+/qysLK15//zzj9Z5E7XRz4N/wGraT/k5K9U9MTk1NRUnT55ETEwMhgwZIrWfP39eq9ba2hpLly7F0qVLkZGRge3bt+P999/H1atXERcXBwDw9vbGli1bIITA77//jpiYGHzwwQdQqVR4//33azz2h302DRo0gKWlJYQQsr/v1tbWUCgUOHDgQIW3cri/rSbbFRoaWq2g07NnT6xYsQLHjh3TOK/J3d1d9jqqUpvvc2162M97bm4uduzYgdmzZ2u810VFRbh582aF63zwd075z9S3334r7UUl/cE9TaQ3zMzM0KlTJ2zbtk3j//Ju376Nn376qVrratSoEV577TW8/fbbuHnzpnTVWfkfl4r+j68mvv76a40b+F26dAlJSUkaV8u5urrizz//1LhS7caNG0hKStJYV3XG5uPjA5VKhU2bNmm0X758GXv37pXujfOoXnzxRQDQ6ufo0aNIS0urUT/NmjWDu7s71q5dK+vqvXLlf1weDA2rVq2qcjlnZ2eMGzcOPXv2xIkTJypcb+vWrfHpp5+iUaNGFdaU8/X1hVqtxhdffKF148ZyHh4eeOaZZ7B582aNmoKCAnz33XfSlV7V+b6HhIRACIG///4bHTp00Jq8vb0fabsaN26stc6qTJw4Eaampnj77bdr9QTzcrX5Ptemh/28KxQKCCG0vqNfffUVSktLZfURFBQEQ0NDXLhwocLPWu7J9/R4cE8T6ZUPP/wQvXr1Qs+ePTF58mSUlpZiwYIFMDMzq/T/1MqFhobCy8sLHTp0gI2NDS5duoSlS5fCxcVFupKs/I/LZ599hiFDhsDIyAgeHh41vpne1atX8eqrryIiIgK5ubmYPXs2TExMEBUVJdWEh4dj1apVGDRoECIiInDjxg0sXLhQ61Jtc3NzuLi44Mcff0T37t1hZWUFa2tr6bYE92vUqBFmzpyJadOmYfDgwRg4cCBu3LiBOXPmwMTEBLNnz67R9jzIw8MDI0eOxLJly9CgQQMEBwdLV885OTlh4sSJNVrvihUrEBoaii5dumDixIlwdnZGRkYGdu/ejdjY2AqXad68Odzd3fH+++9DCAErKyv89NNPSEhI0KjLzc1Ft27dEBYWhubNm8Pc3BxHjx5FXFycdDXUjh07sHLlSrzyyit47rnnIITAtm3bcOvWLfTs2bPScTds2BBLlizBiBEj0KNHD0RERMDOzg7nz5/HyZMnsXz5cjRo0AALFy7Em2++iZCQEIwaNQpFRUVYtGgRbt26hY8//lhan9zvu5+fH0aOHIm33noLx44dQ9euXWFmZoasrCwcPHgQ3t7eGDNmTI23q7rc3d3x9ddfY+DAgVLf5Te3vHr1qnQVZE1vCFvb73NtedjPu4WFBbp27YpFixZJP7v79+/HmjVrZN9M19XVFR988AGmT5+Ov/76C7169YKlpSWuXLmCI0eOwMzMTLoVCulA3Z97TlS17du3i1atWgljY2Ph7OwsPv74Y+nql/s9eEXbkiVLhK+vr7C2tpaWHT58uEhPT9dYLioqSjg6OooGDRoIAOLnn3+W1te7d+8Kx1TZ1XMbN24UkZGRwsbGRiiVSvHCCy+IY8eOaS2/fv164enpKUxMTESLFi3E1q1bta6eE0KIxMRE0bZtW6FUKgUAqc/KriT76quvpPdKrVaLl19+WZw+fVqjZsiQIcLMzExrTBW9pxUpLS0VCxYsEM2aNRNGRkbC2tpaDBo0SGRmZla4PjlXzwnx7xVhwcHBQq1WC6VSKdzd3cXEiROl+RVt85kzZ0TPnj2Fubm5sLS0FK+//rrIyMjQuOrw7t27YvTo0aJVq1bCwsJCqFQq4eHhIWbPni0KCgqEEEL88ccfYuDAgcLd3V2oVCqhVqtFp06dRExMjKyx79q1S/j7+wszMzNhamoqWrRoIRYsWKBR88MPP4jOnTsLExMTYWZmJrp37y5+/fVXrXXJ/b4LIcTatWtF586dhZmZmVCpVMLd3V0MHjxY+s496nZV14ULF8T48eOFh4eHUKlUQqlUChcXF/H666+L77//XuNKs+pcPVeutt7nyr6blf1s+Pv7i5YtW2qNXc7P++XLl0W/fv2EpaWlMDc3F7169RKpqalav0Oq2u7y7erWrZuwsLCQ3tfXXntNJCYmVlhPdUMhRCX7PomIiAj79u1Dt27d8N///hevvfaarodDOsRzmoiIiIhkYGgiIiIikoGH54iIiIhk4J4mIiIiIhkYmoj0QHp6OhQKhTSV35ive/fuD32I7cPW27t3b1hZWUGhUGDChAmV1rq6umqM4f4pPz+/xmOoLUlJSYiOjpYewKwLQ4cORcOGDWu8/O7duxEYGAhHR0colUo4OjoiICCgysvj+/btC4VCgXHjxtW439qyb98+KBSKGj/KIyYmBgqFQrpvWnXs2rUL0dHRNeqXqLYwNBHpkfHjxyM5ORkHDhzA4sWLce7cObz00kv45ZdfarS+iRMn4vDhw1i7di2Sk5Mfel8lPz8/JCcna021fZPAmkhKSsKcOXN0GpoexRdffIFevXrBwsICy5cvx+7du7FgwQJ4enri22+/rXCZq1evYseOHQCA2NjYhz725km2a9cu3p+IdI43tyTSI87OzujSpQuAfwNM06ZN4e/vjzVr1qBr167VXl9qaio6deqEV155RVZ9o0aNpP7luHPnjl4EqooUFhZCpVLpehiS+fPno2vXrloBKTw8HGVlZRUus2HDBpSUlKB3797YuXMntm3bhrCwsIf2VVhYCBMTk2o/FoiIqsY9TUR6rPyRCeUP8SyXnZ2NUaNG4dlnn4WxsTHc3NwwZ84c6en05YdRzp8/j//973/SYbaaHBYpFxAQAC8vL/zyyy/w9fWFqakphg0bBgDIy8vDlClT4ObmBmNjYzzzzDOYMGECCgoKNNYhhMDKlSvRpk0bqFQqWFpa4rXXXsNff/1VZd/R0dF49913AQBubm7S9pQfJnJ1dUVISAi2bduGtm3bwsTERNorsWLFCnTt2hW2trYwMzODt7c3Fi5cWOEDbOPi4tC9e3eo1WqYmprC09MT8+fPr3Jsv/76K6ytrRESEqK1vfe7ceNGpQ+EbtCg4l/Fa9euhZ2dHdavXw+VSoW1a9dq1ZQf8oqPj8ewYcNgY2MDU1NTHDx4EAqFAl9//bXWMhs2bIBCoZCeN3fs2DEMGDAArq6uUKlUcHV1xcCBAyt8KK9chw4dgp+fH0xMTODo6IioqKgK3/OtW7ciMDAQDg4OUKlU8PT0xPvvv6/xXg4dOhQrVqwAAI3DxuXf5+p8xkSPgnuaiPTYxYsXAfz7vLZy2dnZ6NSpExo0aIBZs2bB3d0dycnJ+Oijj5Ceno5169ahXbt2SE5Oxquvvgp3d3csXrwYQMVPhL+fEEIKXuUaNGgg/VHPysrCoEGDMHXqVMybNw8NGjTAnTt34O/vj8uXL2PatGlo1aoVTp8+jVmzZuHUqVNITEyU9niMGjUKMTExiIyMxIIFC3Dz5k188MEH8PX1xcmTJ2FnZ1fhuEaMGIGbN29i2bJl2LZtm7QdLVq0kGpOnDiBtLQ0zJgxA25ubjAzMwMAXLhwAWFhYVKgO3nyJObOnYs//vhDI4SsWbMGERER8Pf3xxdffAFbW1v8+eefSE1NrfT9+uabbzB48GAMGzYMy5Ytg4GBQaW1Pj4++O677xAdHY1XX30VXl5eVdYnJSUhLS0N7777Lho3box+/fohNjYWFy9ehJubm1b9sGHD0Lt3b2zcuBEFBQXw9fVF27ZtsWLFCgwcOFCjdvny5ejYsSM6duwI4N9z3zw8PDBgwABYWVkhKysLn3/+OTp27IgzZ85U+8HMZ86cQffu3eHq6oqYmBiYmppi5cqV2Lx5s1Zt+SHoCRMmwMzMDH/88QcWLFiAI0eOYO/evQCAmTNnoqCgAN9++y2Sk5OlZcu/B3I/Y6JHpsvbkRPRvy5evCgAiAULFoiSkhJx9+5dkZKSInx8fISDg4PGo0RGjRolGjZsKC5duqSxjsWLFwsAGo9RqerRMA9ycXERALSm6dOnCyH+fbQEALFnzx6N5ebPny8aNGig9TiIb7/9VgAQu3btEkL8+9gUAGLJkiUadZmZmUKlUompU6dWOb5FixZV+CiZ8rEbGBiIs2fPVrmO0tJSUVJSIjZs2CAMDAzEzZs3hRBC3L59W1hYWIjnn39e49EfD7r/sRsff/yxMDAw0HqsR2XOnz8vvLy8pPdVpVKJ7t27i+XLl4vi4mKt+mHDhgkAIi0tTQjxf4/ymDlzpkZd+eM4Bg8erLWO8nm//fab1HbkyBEBQKxfv77Ssd67d0/k5+cLMzMz8dlnn0nt5WMof/RQZd544w2hUqlEdna2xjqbN29e6WcohBBlZWWipKRE7N+/XwAQJ0+elOa9/fbbsh/7U9FnTFQbGJqI9EB5aHpwMjc313q21TPPPCNCQ0NFSUmJxnT69GkBQKxcuVKqrW5oev7558XRo0c1pr///lsI8W9osrS01FrOz89PtGrVSms8t2/fFgqFQgpD06dPFwqFQly5ckWrtkuXLqJTp05Vju9hoalt27YVLnfixAkRGhoqrKystN7fQ4cOCSGE2L17twAgNm/eXOUYhgwZIkxNTcXIkSOFUqkUW7durbL+QaWlpWL//v1izpw5IjQ0VFhYWAgAon379qKwsFCqu337tmjYsKHw9fWV2srKyoS7u7twcnISpaWlUnt5MPrxxx+1+rt7966wtbUVI0aMkNrCw8OFjY2NuHv3rkZ/U6dOFe7u7sLAwEDjPRo9erRUJzc02draipCQEK328mfA3f8ZXrhwQQwcOFDY2dkJhUKh0feWLVukuqpCk5zPmKg28PAckR555513MGjQIBQVFeHQoUOYMWMGXn75ZZw8eRKNGzcG8O/5TT/99BOMjIwqXMf169dr3L9arZbOo6pIRYf3rly5gvPnzz90PFeuXIEQotJDcM8991wNRlz12DIyMvDCCy/Aw8MDn332GVxdXWFiYoIjR47g7bffRmFhIQDg2rVrAIBnn332of0UFxdj69ataNmyJYKDg6s1xgYNGqBr167SSf0FBQUYPnw4tm7dirVr12Ls2LEA/j3PJz8/H/3799e4WrB///6YP38+EhISEBQU9NDtVyqVGDVqFJYsWYJFixahpKQE33zzDSZNmgSlUinVhYWFYc+ePZg5cyY6duwICwsLKBQKvPTSS9J7VB03btyAvb29VvuDbfn5+XjhhRdgYmKCjz76CM2aNYOpqSkyMzPRt29fWX3L/YyJagNDE5EeefbZZ6XQ4ufnB3t7ewwaNAizZ8/G8uXLAQDW1tZo1aoV5s6dW+E6HB0dH9v4Kroay9rautKTlMvnl/9XoVDgwIEDGn+wy1XU9qhj++GHH1BQUIBt27bBxcVFak9JSdGos7GxAQBcvnz5of0olUr8/PPPCAoKQo8ePRAXFwdLS8sajdnMzAxRUVHYunWrxrlTa9asAQBMmDChwntrrVmzRis0VXal3JgxY/Dxxx9j7dq1uHv3Lu7du4fRo0dL83Nzc7Fjxw7Mnj0b77//vtReVFSEmzdv1mi7GjdujOzsbK32B9v27t2Lf/75B/v27YO/v7/UXp3bSsj9jIlqA0MTkR5788038dVXX2H16tV499134eLigpCQEOzatQvu7u41/mNdm0JCQjBv3jw0bty4whOU76/7+OOP8ffff6N///7V7qc8VFVnz0F5kLg/kAkhsHr1ao06X19fqNVqfPHFFxgwYMBDL9Vv27Yt9u/fjx49eiAgIAAJCQmwtbWtcpmsrKwK9walpaUB+L+wm5aWhuTkZPTr16/CG1p+9NFH+PHHH3Hjxg1p72NVHBwc8Prrr2PlypUoLi5GaGgonJ2dpfkKhQJCCK3Q+tVXX6G0tPSh669It27dsH37dly5ckXas1haWoqtW7dq1FX0+QDAqlWrtNZ5/+d//60k5H7GRLWBoYlIzy1YsACdO3fGhx9+iK+++goffPABEhIS4Ovri8jISHh4eODu3btIT0/Hrl278MUXX8g6zFRbJkyYgO+++w5du3bFxIkT0apVK5SVlSEjIwPx8fGYPHkyOnfuDD8/P4wcORJvvfUWjh07hq5du8LMzAxZWVk4ePAgvL29MWbMmEr78fb2BgB89tlnGDJkCIyMjODh4QFzc/NKl+nZsyeMjY0xcOBATJ06FXfv3sXnn3+OnJwcjbqGDRtiyZIlGDFiBHr06IGIiAjY2dnh/PnzOHnypLSX736enp44cOAAevToga5duyIxMbHK971ly5bo3r07goOD4e7ujrt37+Lw4cNYsmQJ7OzsMHz4cAD/t5dp6tSp6NSpk9Z6bt++jT179mDTpk145513Ku3vfu+88w46d+4MAFi3bp3GPAsLC3Tt2hWLFi2CtbU1XF1dsX//fqxZswaNGjWStf4HzZgxA9u3b8eLL76IWbNmwdTUFCtWrNC6JYOvry8sLS0xevRozJ49G0ZGRoiNjcXJkye11ln++S9YsADBwcEwMDBAq1atZH/GRLVCt6dUEZEQ/3ci+KJFiyqc//rrrwtDQ0Nx/vx5IYQQ165dE5GRkcLNzU0YGRkJKysr0b59ezF9+nSRn58vLVfdE8GrqvX39xctW7ascF5+fr6YMWOG8PDwEMbGxkKtVgtvb28xceJEjSuohBBi7dq1onPnzsLMzEyoVCrh7u4uBg8erHXCe0WioqKEo6OjaNCggcYJyVWN/aeffhKtW7cWJiYm4plnnhHvvvuu+N///lfhCc27du0S/v7+wszMTJiamooWLVpoXB13/9Vz5S5fviyaN28uXF1dxYULFyod+6pVq0Tfvn3Fc889J0xNTYWxsbFwd3cXo0ePFpmZmUIIIYqLi4Wtra1o06ZNpeu5d++eePbZZ4W3t7cQ4v9OBH/w6sUHubq6Ck9PzwrnXb58WfTr109YWloKc3Nz0atXL5GamipcXFzEkCFDpDq5J4ILIcSvv/4qunTpIpRKpbC3txfvvvuu+PLLL7VOBE9KShI+Pj7C1NRU2NjYiBEjRogTJ04IAGLdunVSXVFRkRgxYoSwsbGRThgvX091PmOiR6EQQgidJTYiInrsfv/9d7Ru3RorVqyQTjYnoupjaCIiekJduHABly5dwrRp05CRkYHz58/r7WNviOoDPkaFiOgJ9eGHH6Jnz57Iz8/Hf//7XwYmokfEPU1EREREMnBPExEREZEMDE1EREREMjA0EREREcnAm1vWorKyMvzzzz8wNzd/6B2FiYiISD8IIXD79m04OjqiQYPK9ycxNNWif/75B05OTroeBhEREdVAZmZmlXf2Z2iqReWPc8jMzISFhYWOR0NERERy5OXlwcnJqcrHMgEMTbWq/JCchYUFQxMREVE987BTa3giOBEREZEMDE1EREREMjA0EREREcnA0EREREQkA0MTERERkQwMTUREREQyMDQRERERycDQRERERCQDQxMRERGRDAxNRERERDIwNBERERHJwNBEREREJANDExEREZEMDE1EREREMhjqegBEcmVkZOD69es669/a2hrOzs4665+IiHSLoYnqhYyMDDRv7onCwjs6G4NKZYo//khjcCIiekoxNFG9cP36dRQW3kHnYbNh4eBa5/3nZaXj8No5uH79OkMTEdFTiqGJ6hULB1dYOXvoehhERPQU4ongRERERDIwNBERERHJwNBEREREJANDExEREZEMDE1EREREMjA0EREREcnA0EREREQkA0MTERERkQwMTUREREQyMDQRERERycDQRERERCQDQxMRERGRDAxNRERERDIwNBERERHJwNBEREREJINOQ9Mvv/yC0NBQODo6QqFQ4IcfftCYr1AoKpwWLVok1QQEBGjNHzBggMZ6cnJyEB4eDrVaDbVajfDwcNy6dUujJiMjA6GhoTAzM4O1tTUiIyNRXFz8uDadiIiI6hmdhqaCggK0bt0ay5cvr3B+VlaWxrR27VooFAr069dPoy4iIkKjbtWqVRrzw8LCkJKSgri4OMTFxSElJQXh4eHS/NLSUvTu3RsFBQU4ePAgtmzZgu+++w6TJ0+u/Y0mIiKieslQl50HBwcjODi40vn29vYar3/88Ud069YNzz33nEa7qampVm25tLQ0xMXF4dChQ+jcuTMAYPXq1fDx8cHZs2fh4eGB+Ph4nDlzBpmZmXB0dAQALFmyBEOHDsXcuXNhYWHxKJtJRERET4B6c07TlStXsHPnTgwfPlxrXmxsLKytrdGyZUtMmTIFt2/fluYlJydDrVZLgQkAunTpArVajaSkJKnGy8tLCkwAEBQUhKKiIhw/frzSMRUVFSEvL09jIiIioieTTvc0Vcf69ethbm6Ovn37arS/+eabcHNzg729PVJTUxEVFYWTJ08iISEBAJCdnQ1bW1ut9dna2iI7O1uqsbOz05hvaWkJY2NjqaYi8+fPx5w5cx5104iIiKgeqDehae3atXjzzTdhYmKi0R4RESH928vLC02bNkWHDh1w4sQJtGvXDsC/J5Q/SAih0S6n5kFRUVGYNGmS9DovLw9OTk7yN4qIiIjqjXpxeO7AgQM4e/YsRowY8dDadu3awcjICOfOnQPw73lRV65c0aq7du2atHfJ3t5ea49STk4OSkpKtPZA3U+pVMLCwkJjIiIioidTvQhNa9asQfv27dG6deuH1p4+fRolJSVwcHAAAPj4+CA3NxdHjhyRag4fPozc3Fz4+vpKNampqcjKypJq4uPjoVQq0b59+1reGiIiIqqPdHp4Lj8/H+fPn5deX7x4ESkpKbCysoKzszOAfw95/fe//8WSJUu0lr9w4QJiY2Px0ksvwdraGmfOnMHkyZPRtm1b+Pn5AQA8PT3Rq1cvRERESLciGDlyJEJCQuDh4QEACAwMRIsWLRAeHo5Fixbh5s2bmDJlCiIiIrj3iIiIiADoeE/TsWPH0LZtW7Rt2xYAMGnSJLRt2xazZs2SarZs2QIhBAYOHKi1vLGxMfbs2YOgoCB4eHggMjISgYGBSExMhIGBgVQXGxsLb29vBAYGIjAwEK1atcLGjRul+QYGBti5cydMTEzg5+eH/v3745VXXsHixYsf49YTERFRfaIQQghdD+JJkZeXB7VajdzcXO6hqmUnTpxA+/bt0XP6Olg5e9R5/zczziJh7ls4fvy4dIEBERE9GeT+/a4X5zQRERER6RpDExEREZEMDE1EREREMjA0EREREclQb+4ITkRPt4yMDFy/fl2nY7C2tpZuh0JETx+GJiLSexkZGWje3BOFhXd0Og6VyhR//JHG4ET0lGJoIiK9d/36dRQW3kHnYbNh4eCqkzHkZaXj8No5uH79OkMT0VOKoYmI6g0LB1ed3KeLiAjgieBEREREsjA0EREREcnA0EREREQkA0MTERERkQwMTUREREQyMDQRERERycDQRERERCQDQxMRERGRDAxNRERERDIwNBERERHJwNBEREREJANDExEREZEMDE1EREREMjA0EREREcnA0EREREQkA0MTERERkQwMTUREREQyMDQRERERycDQRERERCQDQxMRERGRDAxNRERERDIwNBERERHJwNBEREREJANDExEREZEMDE1EREREMjA0EREREcnA0EREREQkA0MTERERkQw6DU2//PILQkND4ejoCIVCgR9++EFj/tChQ6FQKDSmLl26aNQUFRVh/PjxsLa2hpmZGfr06YPLly9r1OTk5CA8PBxqtRpqtRrh4eG4deuWRk1GRgZCQ0NhZmYGa2trREZGori4+HFsNhEREdVDOg1NBQUFaN26NZYvX15pTa9evZCVlSVNu3bt0pg/YcIEfP/999iyZQsOHjyI/Px8hISEoLS0VKoJCwtDSkoK4uLiEBcXh5SUFISHh0vzS0tL0bt3bxQUFODgwYPYsmULvvvuO0yePLn2N5qIiIjqJUNddh4cHIzg4OAqa5RKJezt7Sucl5ubizVr1mDjxo3o0aMHAGDTpk1wcnJCYmIigoKCkJaWhri4OBw6dAidO3cGAKxevRo+Pj44e/YsPDw8EB8fjzNnziAzMxOOjo4AgCVLlmDo0KGYO3cuLCwsanGriYiIqD7S+3Oa9u3bB1tbWzRr1gwRERG4evWqNO/48eMoKSlBYGCg1Obo6AgvLy8kJSUBAJKTk6FWq6XABABdunSBWq3WqPHy8pICEwAEBQWhqKgIx48fr3RsRUVFyMvL05iIiIjoyaTXoSk4OBixsbHYu3cvlixZgqNHj+LFF19EUVERACA7OxvGxsawtLTUWM7Ozg7Z2dlSja2trda6bW1tNWrs7Ow05ltaWsLY2Fiqqcj8+fOl86TUajWcnJweaXuJiIhIf+n08NzDvPHGG9K/vby80KFDB7i4uGDnzp3o27dvpcsJIaBQKKTX9//7UWoeFBUVhUmTJkmv8/LyGJyIiIieUHq9p+lBDg4OcHFxwblz5wAA9vb2KC4uRk5Ojkbd1atXpT1H9vb2uHLlita6rl27plHz4B6lnJwclJSUaO2Bup9SqYSFhYXGRERERE+mehWabty4gczMTDg4OAAA2rdvDyMjIyQkJEg1WVlZSE1Nha+vLwDAx8cHubm5OHLkiFRz+PBh5ObmatSkpqYiKytLqomPj4dSqUT79u3rYtOIiIhIz+n08Fx+fj7Onz8vvb548SJSUlJgZWUFKysrREdHo1+/fnBwcEB6ejqmTZsGa2trvPrqqwAAtVqN4cOHY/LkyWjcuDGsrKwwZcoUeHt7S1fTeXp6olevXoiIiMCqVasAACNHjkRISAg8PDwAAIGBgWjRogXCw8OxaNEi3Lx5E1OmTEFERAT3HhEREREAHYemY8eOoVu3btLr8vODhgwZgs8//xynTp3Chg0bcOvWLTg4OKBbt27YunUrzM3NpWU+/fRTGBoaon///igsLET37t0RExMDAwMDqSY2NhaRkZHSVXZ9+vTRuDeUgYEBdu7cibFjx8LPzw8qlQphYWFYvHjx434LiIiIqJ7QaWgKCAiAEKLS+bt3737oOkxMTLBs2TIsW7as0horKyts2rSpyvU4Oztjx44dD+2PiIiInk716pwmIiIiIl1haCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGXQamn755ReEhobC0dERCoUCP/zwgzSvpKQE7733Hry9vWFmZgZHR0cMHjwY//zzj8Y6AgICoFAoNKYBAwZo1OTk5CA8PBxqtRpqtRrh4eG4deuWRk1GRgZCQ0NhZmYGa2trREZGori4+HFtOhEREdUzOg1NBQUFaN26NZYvX641786dOzhx4gRmzpyJEydOYNu2bfjzzz/Rp08frdqIiAhkZWVJ06pVqzTmh4WFISUlBXFxcYiLi0NKSgrCw8Ol+aWlpejduzcKCgpw8OBBbNmyBd999x0mT55c+xtNRERE9ZKhLjsPDg5GcHBwhfPUajUSEhI02pYtW4ZOnTohIyMDzs7OUrupqSns7e0rXE9aWhri4uJw6NAhdO7cGQCwevVq+Pj44OzZs/Dw8EB8fDzOnDmDzMxMODo6AgCWLFmCoUOHYu7cubCwsKiNzSUiIqJ6rF6d05SbmwuFQoFGjRpptMfGxsLa2hotW7bElClTcPv2bWlecnIy1Gq1FJgAoEuXLlCr1UhKSpJqvLy8pMAEAEFBQSgqKsLx48crHU9RURHy8vI0JiIiInoy6XRPU3XcvXsX77//PsLCwjT2/Lz55ptwc3ODvb09UlNTERUVhZMnT0p7qbKzs2Fra6u1PltbW2RnZ0s1dnZ2GvMtLS1hbGws1VRk/vz5mDNnTm1sHhEREem5ehGaSkpKMGDAAJSVlWHlypUa8yIiIqR/e3l5oWnTpujQoQNOnDiBdu3aAQAUCoXWOoUQGu1yah4UFRWFSZMmSa/z8vLg5OQkf8OIiIio3tD7w3MlJSXo378/Ll68iISEhIeeX9SuXTsYGRnh3LlzAAB7e3tcuXJFq+7atWvS3iV7e3utPUo5OTkoKSnR2gN1P6VSCQsLC42JiIiInkx6HZrKA9O5c+eQmJiIxo0bP3SZ06dPo6SkBA4ODgAAHx8f5Obm4siRI1LN4cOHkZubC19fX6kmNTUVWVlZUk18fDyUSiXat29fy1tFRERE9ZFOD8/l5+fj/Pnz0uuLFy8iJSUFVlZWcHR0xGuvvYYTJ05gx44dKC0tlfYGWVlZwdjYGBcuXEBsbCxeeuklWFtb48yZM5g8eTLatm0LPz8/AICnpyd69eqFiIgI6VYEI0eOREhICDw8PAAAgYGBaNGiBcLDw7Fo0SLcvHkTU6ZMQUREBPceEREREQAd72k6duwY2rZti7Zt2wIAJk2ahLZt22LWrFm4fPkytm/fjsuXL6NNmzZwcHCQpvKr3oyNjbFnzx4EBQXBw8MDkZGRCAwMRGJiIgwMDKR+YmNj4e3tjcDAQAQGBqJVq1bYuHGjNN/AwAA7d+6EiYkJ/Pz80L9/f7zyyitYvHhx3b4hREREpLd0uqcpICAAQohK51c1DwCcnJywf//+h/ZjZWWFTZs2VVnj7OyMHTt2PHRdRERE9HTS63OaiIiIiPQFQxMRERGRDAxNRERERDIwNBERERHJwNBEREREJANDExEREZEMDE1EREREMjA0EREREcnA0EREREQkA0MTERERkQwMTUREREQyMDQRERERycDQRERERCQDQxMRERGRDAxNRERERDIwNBERERHJwNBEREREJANDExEREZEMDE1EREREMjA0EREREcnA0EREREQkA0MTERERkQwMTUREREQyMDQRERERycDQRERERCQDQxMRERGRDDUKTc899xxu3Lih1X7r1i0899xzjzwoIiIiIn1To9CUnp6O0tJSrfaioiL8/fffjzwoIiIiIn1jWJ3i7du3S//evXs31Gq19Lq0tBR79uyBq6trrQ2OiIiISF9UKzS98sorAACFQoEhQ4ZozDMyMoKrqyuWLFlSa4MjIiIi0hfVCk1lZWUAADc3Nxw9ehTW1taPZVBERERE+qZaoancxYsXa3scRERERHqtRqEJAPbs2YM9e/bg6tWr0h6ocmvXrn3kgRERERHpkxqFpjlz5uCDDz5Ahw4d4ODgAIVCUdvjIiIiItIrNQpNX3zxBWJiYhAeHl7b4yEiIiLSSzW6T1NxcTF8fX1reyxEREREeqtGoWnEiBHYvHnzI3f+yy+/IDQ0FI6OjlAoFPjhhx805gshEB0dDUdHR6hUKgQEBOD06dMaNUVFRRg/fjysra1hZmaGPn364PLlyxo1OTk5CA8Ph1qthlqtRnh4OG7duqVRk5GRgdDQUJiZmcHa2hqRkZEoLi5+5G0kIiKiJ0ONDs/dvXsXX375JRITE9GqVSsYGRlpzP/kk09kraegoACtW7fGW2+9hX79+mnNX7hwIT755BPExMSgWbNm+Oijj9CzZ0+cPXsW5ubmAIAJEybgp59+wpYtW9C4cWNMnjwZISEhOH78OAwMDAAAYWFhuHz5MuLi4gAAI0eORHh4OH766ScA/96Ys3fv3rCxscHBgwdx48YNDBkyBEIILFu2rCZvERERET1hahSafv/9d7Rp0wYAkJqaqjGvOieFBwcHIzg4uMJ5QggsXboU06dPR9++fQEA69evh52dHTZv3oxRo0YhNzcXa9aswcaNG9GjRw8AwKZNm+Dk5ITExEQEBQUhLS0NcXFxOHToEDp37gwAWL16NXx8fHD27Fl4eHggPj4eZ86cQWZmJhwdHQEAS5YswdChQzF37lxYWFhU6/0hIiKiJ0+NQtPPP/9c2+PQcvHiRWRnZyMwMFBqUyqV8Pf3R1JSEkaNGoXjx4+jpKREo8bR0RFeXl5ISkpCUFAQkpOToVarpcAEAF26dIFarUZSUhI8PDyQnJwMLy8vKTABQFBQEIqKinD8+HF069atwjEWFRWhqKhIep2Xl1ebbwERERHpkRqd01QXsrOzAQB2dnYa7XZ2dtK87OxsGBsbw9LSssoaW1tbrfXb2tpq1DzYj6WlJYyNjaWaisyfP186T0qtVsPJyamaW0lERET1RY32NHXr1q3Kw3B79+6t8YAe9GA/QoiHHgJ8sKai+prUPCgqKgqTJk2SXufl5TE4ERERPaFqFJrKz2cqV1JSgpSUFKSmpmo9yLem7O3tAfy7F8jBwUFqv3r1qrRXyN7eHsXFxcjJydHY23T16lXplgj29va4cuWK1vqvXbumsZ7Dhw9rzM/JyUFJSYnWHqj7KZVKKJXKGm4hERER1Sc1Ck2ffvpphe3R0dHIz89/pAGVc3Nzg729PRISEtC2bVsA/94fav/+/ViwYAEAoH379jAyMkJCQgL69+8PAMjKykJqaioWLlwIAPDx8UFubi6OHDmCTp06AQAOHz6M3NxcKVj5+Phg7ty5yMrKkgJafHw8lEol2rdvXyvbQ0RERPVbjZ89V5FBgwahU6dOWLx4saz6/Px8nD9/Xnp98eJFpKSkwMrKCs7OzpgwYQLmzZuHpk2bomnTppg3bx5MTU0RFhYGAFCr1Rg+fDgmT56Mxo0bw8rKClOmTIG3t7d0NZ2npyd69eqFiIgIrFq1CsC/txwICQmBh4cHACAwMBAtWrRAeHg4Fi1ahJs3b2LKlCmIiIjglXNEREQEoJZDU3JyMkxMTGTXHzt2TOPKtPLzg4YMGYKYmBhMnToVhYWFGDt2LHJyctC5c2fEx8dL92gC/t3rZWhoiP79+6OwsBDdu3dHTEyMdI8mAIiNjUVkZKR0lV2fPn2wfPlyab6BgQF27tyJsWPHws/PDyqVCmFhYbLDHxERET35ahSayu+bVE4IgaysLBw7dgwzZ86UvZ6AgAAIISqdr1AoEB0djejo6EprTExMsGzZsipvQmllZYVNmzZVORZnZ2fs2LHjoWMmIiKip1ONQpNardZ43aBBA3h4eOCDDz7QuGcSERER0ZOiRqFp3bp1tT0OIiIiIr32SOc0HT9+HGlpaVAoFGjRooV0lRsRERHRk6ZGoenq1asYMGAA9u3bh0aNGkEIgdzcXHTr1g1btmyBjY1NbY+TiIiISKdq9BiV8ePHIy8vD6dPn8bNmzeRk5OD1NRU5OXlITIysrbHSERERKRzNdrTFBcXh8TERHh6ekptLVq0wIoVK3giOBERET2RarSnqaysDEZGRlrtRkZGKCsre+RBEREREembGoWmF198Ee+88w7++ecfqe3vv//GxIkT0b1791obHBEREZG+qFFoWr58OW7fvg1XV1e4u7ujSZMmcHNzw+3bt6u8ySQRERFRfVWjc5qcnJxw4sQJJCQk4I8//oAQAi1atJCe90ZERET0pKnWnqa9e/eiRYsWyMvLAwD07NkT48ePR2RkJDp27IiWLVviwIEDj2WgRERERLpUrdC0dOlSREREwMLCQmueWq3GqFGj8Mknn9Ta4IiIiIj0RbVC08mTJ9GrV69K5wcGBuL48eOPPCgiIiIifVOt0HTlypUKbzVQztDQENeuXXvkQRERERHpm2qFpmeeeQanTp2qdP7vv/8OBweHRx4UERERkb6pVmh66aWXMGvWLNy9e1drXmFhIWbPno2QkJBaGxwRERGRvqjWLQdmzJiBbdu2oVmzZhg3bhw8PDygUCiQlpaGFStWoLS0FNOnT39cYyUiIiLSmWqFJjs7OyQlJWHMmDGIioqCEAIAoFAoEBQUhJUrV8LOzu6xDJSIiIhIl6p9c0sXFxfs2rULOTk5OH/+PIQQaNq0KSwtLR/H+IiIiIj0Qo3uCA4AlpaW6NixY22OhYiIiEhv1ejZc0RERERPG4YmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikkHvQ5OrqysUCoXW9PbbbwMAhg4dqjWvS5cuGusoKirC+PHjYW1tDTMzM/Tp0weXL1/WqMnJyUF4eDjUajXUajXCw8Nx69atutpMIiIi0nN6H5qOHj2KrKwsaUpISAAAvP7661JNr169NGp27dqlsY4JEybg+++/x5YtW3Dw4EHk5+cjJCQEpaWlUk1YWBhSUlIQFxeHuLg4pKSkIDw8vG42koiIiPSeoa4H8DA2NjYarz/++GO4u7vD399falMqlbC3t69w+dzcXKxZswYbN25Ejx49AACbNm2Ck5MTEhMTERQUhLS0NMTFxeHQoUPo3LkzAGD16tXw8fHB2bNn4eHh8Zi2joiIiOoLvd/TdL/i4mJs2rQJw4YNg0KhkNr37dsHW1tbNGvWDBEREbh69ao07/jx4ygpKUFgYKDU5ujoCC8vLyQlJQEAkpOToVarpcAEAF26dIFarZZqKlJUVIS8vDyNiYiIiJ5M9So0/fDDD7h16xaGDh0qtQUHByM2NhZ79+7FkiVLcPToUbz44osoKioCAGRnZ8PY2BiWlpYa67Kzs0N2drZUY2trq9Wfra2tVFOR+fPnS+dAqdVqODk51cJWEhERkT7S+8Nz91uzZg2Cg4Ph6Ogotb3xxhvSv728vNChQwe4uLhg586d6Nu3b6XrEkJo7K26/9+V1TwoKioKkyZNkl7n5eUxOBERET2h6k1ounTpEhITE7Ft27Yq6xwcHODi4oJz584BAOzt7VFcXIycnByNvU1Xr16Fr6+vVHPlyhWtdV27dg12dnaV9qVUKqFUKmuyOURERFTP1JvDc+vWrYOtrS169+5dZd2NGzeQmZkJBwcHAED79u1hZGQkXXUHAFlZWUhNTZVCk4+PD3Jzc3HkyBGp5vDhw8jNzZVqiIiI6OlWL/Y0lZWVYd26dRgyZAgMDf9vyPn5+YiOjka/fv3g4OCA9PR0TJs2DdbW1nj11VcBAGq1GsOHD8fkyZPRuHFjWFlZYcqUKfD29paupvP09ESvXr0QERGBVatWAQBGjhyJkJAQXjlHREREAOpJaEpMTERGRgaGDRum0W5gYIBTp05hw4YNuHXrFhwcHNCtWzds3boV5ubmUt2nn34KQ0ND9O/fH4WFhejevTtiYmJgYGAg1cTGxiIyMlK6yq5Pnz5Yvnx53WwgERER6b16EZoCAwMhhNBqV6lU2L1790OXNzExwbJly7Bs2bJKa6ysrLBp06ZHGicRERE9uerNOU1EREREusTQRERERCQDQxMRERGRDAxNRERERDIwNBERERHJwNBEREREJANDExEREZEMDE1EREREMjA0EREREcnA0EREREQkQ714jAoRARkZGbh+/brO+re2toazs7PO+ici0jWGJqJ6ICMjA82be6Kw8I7OxqBSmeKPP9IYnIjoqcXQRFQPXL9+HYWFd9B52GxYOLjWef95Wek4vHYODhw4AE9PzzrvPy0trc77JCJ6EEMTUT1i4eAKK2ePOu+3MPcGAAUGDRpU533fr6SoWKf9E9HTjaGJiB6q5M5tAAJtwt6DjVvzOu8/61QyUrd/iXv37tV530RE5RiaiEi2hrbOOtnTlZeVXud9EhE9iLccICIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSQa9DU3R0NBQKhcZkb28vzRdCIDo6Go6OjlCpVAgICMDp06c11lFUVITx48fD2toaZmZm6NOnDy5fvqxRk5OTg/DwcKjVaqjVaoSHh+PWrVt1sYlERERUT+h1aAKAli1bIisrS5pOnTolzVu4cCE++eQTLF++HEePHoW9vT169uyJ27dvSzUTJkzA999/jy1btuDgwYPIz89HSEgISktLpZqwsDCkpKQgLi4OcXFxSElJQXh4eJ1uJxEREek3Q10P4GEMDQ019i6VE0Jg6dKlmD59Ovr27QsAWL9+Pezs7LB582aMGjUKubm5WLNmDTZu3IgePXoAADZt2gQnJyckJiYiKCgIaWlpiIuLw6FDh9C5c2cAwOrVq+Hj44OzZ8/Cw8Oj0rEVFRWhqKhIep2Xl1ebm05ERER6RO/3NJ07dw6Ojo5wc3PDgAED8NdffwEALl68iOzsbAQGBkq1SqUS/v7+SEpKAgAcP34cJSUlGjWOjo7w8vKSapKTk6FWq6XABABdunSBWq2Waiozf/586ZCeWq2Gk5NTrW03ERER6Re9Dk2dO3fGhg0bsHv3bqxevRrZ2dnw9fXFjRs3kJ2dDQCws7PTWMbOzk6al52dDWNjY1haWlZZY2trq9W3ra2tVFOZqKgo5ObmSlNmZmaNt5WIiIj0m14fngsODpb+7e3tDR8fH7i7u2P9+vXo0qULAEChUGgsI4TQanvQgzUV1ctZj1KphFKpfOh2EBERUf2n13uaHmRmZgZvb2+cO3dOOs/pwb1BV69elfY+2dvbo7i4GDk5OVXWXLlyRauva9euae3FIiIioqdXvQpNRUVFSEtLg4ODA9zc3GBvb4+EhARpfnFxMfbv3w9fX18AQPv27WFkZKRRk5WVhdTUVKnGx8cHubm5OHLkiFRz+PBh5ObmSjVEREREen14bsqUKQgNDYWzszOuXr2Kjz76CHl5eRgyZAgUCgUmTJiAefPmoWnTpmjatCnmzZsHU1NThIWFAQDUajWGDx+OyZMno3HjxrCyssKUKVPg7e0tXU3n6emJXr16ISIiAqtWrQIAjBw5EiEhIVVeOUdERERPF70OTZcvX8bAgQNx/fp12NjYoEuXLjh06BBcXFwAAFOnTkVhYSHGjh2LnJwcdO7cGfHx8TA3N5fW8emnn8LQ0BD9+/dHYWEhunfvjpiYGBgYGEg1sbGxiIyMlK6y69OnD5YvX163G0tERER6Ta9D05YtW6qcr1AoEB0djejo6EprTExMsGzZMixbtqzSGisrK2zatKmmwyQiIqKnQL06p4mIiIhIVxiaiIiIiGRgaCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBoYmIiIiIhkYmoiIiIhkYGgiIiIikoGhiYiIiEgGhiYiIiIiGRiaiIiIiGRgaCIiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpLBUNcDIHkyMjJw/fp1nfVvbW0NZ2dnnfVPRESkawxN9UBGRgaaN/dEYeEdnY1BpTLFH3+kMTgREdFTi6GpHrh+/ToKC++g87DZsHBwrfP+87LScXjtHFy/fp2hiYiInloMTfWIhYMrrJw9dD0MIiKipxJPBCciIiKSQa9D0/z589GxY0eYm5vD1tYWr7zyCs6ePatRM3ToUCgUCo2pS5cuGjVFRUUYP348rK2tYWZmhj59+uDy5csaNTk5OQgPD4darYZarUZ4eDhu3br1uDeRiIiI6gm9Dk379+/H22+/jUOHDiEhIQH37t1DYGAgCgoKNOp69eqFrKwsadq1a5fG/AkTJuD777/Hli1bcPDgQeTn5yMkJASlpaVSTVhYGFJSUhAXF4e4uDikpKQgPDy8TraTiIiI9J9en9MUFxen8XrdunWwtbXF8ePH0bVrV6ldqVTC3t6+wnXk5uZizZo12LhxI3r06AEA2LRpE5ycnJCYmIigoCCkpaUhLi4Ohw4dQufOnQEAq1evho+PD86ePQsPj4rPIyoqKkJRUZH0Oi8v75G2l4iIiPSXXu9pelBubi4AwMrKSqN93759sLW1RbNmzRAREYGrV69K844fP46SkhIEBgZKbY6OjvDy8kJSUhIAIDk5GWq1WgpMANClSxeo1WqppiLz58+XDuep1Wo4OTnVynYSERGR/qk3oUkIgUmTJuH555+Hl5eX1B4cHIzY2Fjs3bsXS5YswdGjR/Hiiy9Ke4Cys7NhbGwMS0tLjfXZ2dkhOztbqrG1tdXq09bWVqqpSFRUFHJzc6UpMzOzNjaViIiI9JBeH56737hx4/D777/j4MGDGu1vvPGG9G8vLy906NABLi4u2LlzJ/r27Vvp+oQQUCgU0uv7/11ZzYOUSiWUSmV1NoOIiIjqqXqxp2n8+PHYvn07fv75Zzz77LNV1jo4OMDFxQXnzp0DANjb26O4uBg5OTkadVevXoWdnZ1Uc+XKFa11Xbt2TaohIiKip5tehyYhBMaNG4dt27Zh7969cHNze+gyN27cQGZmJhwcHAAA7du3h5GRERISEqSarKwspKamwtfXFwDg4+OD3NxcHDlyRKo5fPgwcnNzpRoiIiJ6uun14bm3334bmzdvxo8//ghzc3Pp/CK1Wg2VSoX8/HxER0ejX79+cHBwQHp6OqZNmwZra2u8+uqrUu3w4cMxefJkNG7cGFZWVpgyZQq8vb2lq+k8PT3Rq1cvREREYNWqVQCAkSNHIiQkpNIr54iIiOjpoteh6fPPPwcABAQEaLSvW7cOQ4cOhYGBAU6dOoUNGzbg1q1bcHBwQLdu3bB161aYm5tL9Z9++ikMDQ3Rv39/FBYWonv37oiJiYGBgYFUExsbi8jISOkquz59+mD58uWPfyOJiIioXtDr0CSEqHK+SqXC7t27H7oeExMTLFu2DMuWLau0xsrKCps2bar2GImIiOjpoNfnNBERERHpC4YmIiIiIhn0+vAc6Ze0tLSnsm8iIiKAoYlkKMy9AUCBQYMG6XooKCkq1vUQiIjoKcXQRA9Vcuc2AIE2Ye/Bxq25TsaQdSoZqdu/xL1793TSPxEREUMTydbQ1hlWzrq5b1VeVrpO+iV6kC4PFVtbW8PZ2Vln/RM97RiaiIhk0IfD1CqVKf74I43BiUhHGJqIiGTQ9WHqvKx0HF47B9evX2doItIRhiYiomrQ5WFqXcvIyMD169d11j8PT5KuMTQREdFDZWRkoHlzTxQW3tHZGHh4knSNoYmIiB7q+vXrKCy8g87DZsPCwbXO++fhSdIHDE1ERCSbhYPrU3t4koiPUSEiIiKSgaGJiIiISAaGJiIiIiIZGJqIiIiIZGBoIiIiIpKBoYmIiIhIBt5ygIiI6g0+MJl0iaGJiIj0Hh+YTPqAoYmIiPQeH5hM+oChiYiI6o2n+YHJpHs8EZyIiIhIBu5pIiKqR3R1IrQuT8Am0hcMTURE9YA+nAgNACVFxTrtn0iXGJqIiOoBXZ8InXUqGanbv8S9e/fqvG8ifcHQRERUj+jqROi8rPQ675NI3zA0ERERyaTrc7t4g03dYmgiIiJ6CH05p4w32NQthiYiIqKH0PU5ZQBvsKkPGJqIqoGXexM93XhzzacbQxORDPqya56XexMR6Q5DE5EMut41z8u9iYh0j6GJqBp4uTcR0dOLz54jIiIikoGhiYiIiEgGhqYHrFy5Em5ubjAxMUH79u1x4MABXQ+JiIiI9ABD0322bt2KCRMmYPr06fjtt9/wwgsvIDg4GBkZGboeGhEREekYTwS/zyeffILhw4djxIgRAIClS5di9+7d+PzzzzF//nwdj46IiEi392172h/jwtD0/xUXF+P48eN4//33NdoDAwORlJRU4TJFRUUoKiqSXufm5gIA8vLyanVs+fn5AICbl87iXlFhra5bjrysSwCA3L/PwchQUef968MY2P/T3b8+jIH9P939A8D1C6cAQKf3i1MqTbBx4wbY2dnppH97e3vY29vX+nrL/24LIaouFCSEEOLvv/8WAMSvv/6q0T537lzRrFmzCpeZPXu2AMCJEydOnDhxegKmzMzMKrMC9zQ9QKHQ/D8IIYRWW7moqChMmjRJel1WVoabN2+icePGlS6jC3l5eXByckJmZiYsLCyeyjGw/6e7f30YA/t/uvvXhzE87f1XRQiB27dvw9HRsco6hqb/z9raGgYGBsjOztZov3r1aqW7IZVKJZRKpUZbo0aNHtcQH5mFhYXOv6i6HgP7f7r714cxsP+nu399GMPT3n9l1Gr1Q2t49dz/Z2xsjPbt2yMhIUGjPSEhAb6+vjoaFREREekL7mm6z6RJkxAeHo4OHTrAx8cHX375JTIyMjB69GhdD42IiIh0jKHpPm+88QZu3LiBDz74AFlZWfDy8sKuXbvg4uKi66E9EqVSidmzZ2sdSnyaxsD+n+7+9WEM7P/p7l8fxvC0918bFEI87Po6IiIiIuI5TUREREQyMDQRERERycDQRERERCQDQxPRUyAgIAATJkx46vq+39ChQ/HKK6/oehhEVI/x6jmip8C2bdtgZGSk62Ho1Gefffbw50oREVWBoYnoKWBlZaXrIeicnLv9EhFVhYfnnmCurq5YunSpRlubNm0QHR1dZ2MoKyvDggUL0KRJEyiVSjg7O2Pu3LmPvd958+ahYcOGVU4HDhx47OMICAjAuHHjMG7cODRq1AiNGzfGjBkz6nyPh74cIgOAuLg4qNVqbNiwoU771cXhuWvXrsHe3h7z5s2T2g4fPgxjY2PEx8fXyRgCAgIQGRmJqVOnwsrKCvb29nXyOyAgIADjx4/HhAkTYGlpCTs7O3z55ZcoKCjAW2+9BXNzc7i7u+N///vfYx9LUVERIiMjYWtrCxMTEzz//PM4evToY++3nBACCxcuxHPPPQeVSoXWrVvj22+/rZO+N2zYgMaNG6OoqEijvV+/fhg8eHCdjAEAvv32W3h7e0OlUqFx48bo0aMHCgoK6qz/2sLQRI9VVFQUFixYgJkzZ+LMmTPYvHlzpc/yq02jR49GSkpKlVOHDh0e+zgAYP369TA0NMThw4fxn//8B59++im++uqrOulb32zZsgX9+/fHhg0b6vQXtq7Y2Nhg7dq1iI6OxrFjx5Cfn49BgwZh7NixCAwMrLNxrF+/HmZmZjh8+DAWLlyIDz74QOuRUY+rX2traxw5cgTjx4/HmDFj8Prrr8PX1xcnTpxAUFAQwsPDcefOncc6jqlTp+K7777D+vXrceLECTRp0gRBQUG4efPmY+233IwZM7Bu3Tp8/vnnOH36NCZOnIhBgwZh//79j73v119/HaWlpdi+fbvUdv36dezYsQNvvfXWY+8fALKysjBw4EAMGzYMaWlp2LdvH/r27Vs/D5cLemK5uLiITz/9VKOtdevWYvbs2XXSf15enlAqlWL16tV10p8+8vf3F56enqKsrExqe++994Snp2edj+Odd96p0z4f7HvFihVCrVaLvXv36mQcQ4YMES+//LJO+h47dqxo1qyZePPNN4WXl5coLCyss779/f3F888/r9HWsWNH8d5779Vpv/fu3RNmZmYiPDxcasvKyhIARHJy8mMbR35+vjAyMhKxsbFSW3FxsXB0dBQLFy58bP3e37+JiYlISkrSaB8+fLgYOHDgY+9fCCHGjBkjgoODpddLly4Vzz33nMbvpcfp+PHjAoBIT0+vk/4eJ+5poscmLS0NRUVF6N69e533rS+H5wCgS5cuUCgU0msfHx+cO3cOpaWlddK/Pvjuu+8wYcIExMfHo1u3broeTp1bvHgx7t27h2+++QaxsbEwMTGp0/5btWql8drBwQFXr16t034NDAzQuHFjeHt7S23le50f51guXLiAkpIS+Pn5SW1GRkbo1KkT0tLSHlu/5c6cOYO7d++iZ8+eGr9/NmzYgAsXLjz2/gEgIiIC8fHx+PvvvwEA69atw9ChQzV+Lz1OrVu3Rvfu3eHt7Y3XX38dq1evRk5OTp30Xdt4IvgTrEGDBlq7P0tKSuqsf5VKVWd9PWj06NHo379/lTXPPPNMHY2G2rRpgxMnTmDdunXo2LFjnf2y1hd//fUX/vnnH5SVleHSpUtaIeZxe/DKSYVCgbKyMp30e39b+ffgcY6l/Hfgg985IUSdfA/Lt23nzp1av3Pq6hlsbdu2RevWrbFhwwYEBQXh1KlT+Omnn+qkb+DfwJyQkICkpCTEx8dj2bJlmD59Og4fPgw3N7c6G0dtYGh6gtnY2CArK0t6nZeXh4sXL9ZZ/02bNoVKpcKePXswYsSIOusX+PdqMX25YuzQoUNar5s2bQoDAwMdjajuubu7Y8mSJQgICICBgQGWL1+u6yHVmeLiYrz55pt444030Lx5cwwfPhynTp2qk3P7CGjSpAmMjY1x8OBBhIWFAfj3fx6PHTtWJxdHtGjRAkqlEhkZGfD393/s/VVmxIgR+PTTT/H333+jR48ecHJyqtP+FQoF/Pz84Ofnh1mzZsHFxQXff/89Jk2aVKfjeFQMTU+wF198ETExMQgNDYWlpSVmzpxZp3+oTUxM8N5772Hq1KkwNjaGn58frl27htOnT2P48OF1Ng5dy8zMxKRJkzBq1CicOHECy5Ytw5IlS3Q9rDrXrFkz/PzzzwgICIChoaHWlZ1PqunTpyM3Nxf/+c9/0LBhQ/zvf//D8OHDsWPHDl0P7algZmaGMWPG4N1334WVlRWcnZ2xcOFC3Llzp05+D5mbm2PKlCmYOHEiysrK8PzzzyMvLw9JSUlo2LAhhgwZ8tjHAABvvvkmpkyZgtWrV9f5lauHDx/Gnj17EBgYCFtbWxw+fBjXrl2Dp6dnnY6jNjA0PcGioqLw119/ISQkBGq1Gh9++GGd7mkCgJkzZ8LQ0BCzZs3CP//8AwcHB4wePbpOx6BrgwcPRmFhITp16gQDAwOMHz8eI0eO1PWwdMLDwwN79+6V9jg96eFx3759WLp0KX7++WdYWFgAADZu3IhWrVrh888/x5gxY3Q8wqfDxx9/jLKyMoSHh+P27dvo0KEDdu/eDUtLyzrp/8MPP4StrS3mz5+Pv/76C40aNUK7du0wbdq0OukfACwsLNCvXz/s3Lmzzm+9YWFhgV9++QVLly5FXl4eXFxcsGTJEgQHB9fpOGqDQjx40gsR1ZqAgAC0adPmqdmros8GDhwIAwMDbNq0SddDIdKJnj17wtPTE//5z390PZR6i1fPEdET7d69ezhz5gySk5PRsmVLXQ+HqM7dvHkTW7Zswd69e/H222/rejj1Gg/PEdETLTU1Fb6+vujWrdtTd2iYCADatWuHnJwcLFiwAB4eHroeTr3Gw3NEREREMvDwHBEREZEMDE1EREREMjA0EREREcnA0EREREQkA0MTERERkQwMTURU76Snp0OhUCAlJUXXQ6l1+/btg0KhwK1bt3Q9FCJ6AEMTERERkQwMTUREREQyMDQRkd4qKyvDggUL0KRJEyiVSjg7O2Pu3LladaWlpRg+fDjc3NygUqng4eGBzz77TKNm37596NSpE8zMzNCoUSP4+fnh0qVLAICTJ0+iW7duMDc3h4WFBdq3b49jx47JGuOvv/4Kf39/mJqawtLSEkFBQcjJyQEAFBUVITIyEra2tjAxMcHzzz+Po0ePaiy/a9cuNGvWDCqVCt26dUN6erpWH0lJSejatStUKhWcnJwQGRmJgoICWeMjotrD0EREeisqKgoLFizAzJkzcebMGWzevBl2dnZadWVlZXj22WfxzTff4MyZM5g1axamTZuGb775BsC/z5975ZVX4O/vj99//x3JyckYOXIkFAoFAODNN9/Es88+i6NHj+L48eN4//33YWRk9NDxpaSkoHv37mjZsiWSk5Nx8OBBhIaGorS0FAAwdepUfPfdd1i/fj1OnDiBJk2aICgoCDdv3gQAZGZmom/fvnjppZeQkpKCESNG4P3339fo49SpUwgKCkLfvn3x+++/Y+vWrTh48CDGjRv3SO8tEdWAICLSQ3l5eUKpVIrVq1drzbt48aIAIH777bdKlx87dqzo16+fEEKIGzduCABi3759Fdaam5uLmJiYao9x4MCBws/Pr8J5+fn5wsjISMTGxkptxcXFwtHRUSxcuFAIIURUVJTw9PQUZWVlUs17770nAIicnBwhhBDh4eFi5MiRGus+cOCAaNCggSgsLKz2mImo5riniYj0UlpaGoqKitC9e3dZ9V988QU6dOgAGxsbNGzYEKtXr0ZGRgYAwMrKCkOHDkVQUBBCQ0Px2WefISsrS1p20qRJGDFiBHr06IGPP/4YFy5ckNVn+Z6mily4cAElJSXw8/OT2oyMjNCpUyekpaVJ29ilSxdpjxcA+Pj4aKzn+PHjiImJQcOGDaUpKCgIZWVluHjxoqxxElHtYGgiIr2kUqlk137zzTeYOHEihg0bhvj4eKSkpOCtt95CcXGxVLNu3TokJyfD19cXW7duRbNmzXDo0CEAQHR0NE6fPo3evXtj7969aNGiBb7//vtHGqP4/89Cvz8QlbeXtwkZz0svKyvDqFGjkJKSIk0nT57EuXPn4O7u/tDliaj2MDQRkV5q2rQpVCoV9uzZ89DaAwcOwNfXF2PHjkXbtm3RpEmTCvcWtW3bFlFRUUhKSoKXlxc2b94szWvWrBkmTpyI+Ph49O3bF+vWrXtov61atap0fE2aNIGxsTEOHjwotZWUlODYsWPw9PQEALRo0UIKbuUefN2uXTucPn0aTZo00ZqMjY0fOkYiqj0MTUSkl0xMTPDee+9h6tSp2LBhAy5cuIBDhw5hzZo1WrVNmjTBsWPHsHv3bvz555+YOXOmxlVqFy9eRFRUFJKTk3Hp0iXEx8fjzz//hKenJwoLCzFu3Djs27cPly5dwq+//oqjR49KwaYqUVFROHr0KMaOHYvff/8df/zxBz7//HNcv34dZmZmGDNmDN59913ExcXhzJkziIiIwJ07dzB8+HAAwOjRo3HhwgVMmjQJZ8+exebNmxETE6PRx3vvvYfk5GS8/fbbSElJwblz57B9+3aMHz/+0d5gIqo+HZ9TRURUqdLSUvHRRx8JFxcXYWRkJJydncW8efO0TgS/e/euGDp0qFCr1aJRo0ZizJgx4v333xetW7cWQgiRnZ0tXnnlFeHg4CCMjY2Fi4uLmDVrligtLRVFRUViwIABwsnJSRgbGwtHR0cxbtw42SdZ79u3T/j6+gqlUikaNWokgoKCpJO4CwsLxfjx44W1tbVQKpXCz89PHDlyRGP5n376STRp0kQolUrxwgsviLVr12qcCC6EEEeOHBE9e/YUDRs2FGZmZqJVq1Zi7ty5j/r2ElE1KYSQcVCdiIiI6CnHw3NEREREMjA0ERFVIjg4WONS//unefPm6Xp4RFTHeHiOiKgSf//9NwoLCyucZ2VlBSsrqzoeERHpEkMTERERkQw8PEdEREQkA0MTERERkQwMTUREREQyMDQRERERycDQRERERCQDQxMRERGRDAxNRERERDL8P8bm937xhr0RAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(tracking_df.class_code).set(title='distribution of class codes - Gffcompare \\n Ref free track test')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16",
   "metadata": {},
   "source": [
    "<h5>We have combined count matrix, generated with Isoquant in Quantification mode, quantifying against a reconstrutced gtf - stringtie_merged.noRef.gtf </h5>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform_id</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.1.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100.1</td>\n",
       "      <td>476.0</td>\n",
       "      <td>596.5</td>\n",
       "      <td>487.5</td>\n",
       "      <td>492.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>479.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>364.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.100.2</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.1000.1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     isoform_id   bc01   bc02   bc03   bc04   bc05   bc06   bc07   bc08  \\\n",
       "0     MSTRG.1.1    0.0    0.0    0.0    1.0    1.0    0.0    0.0    0.0   \n",
       "1    MSTRG.10.1    0.0    0.0    1.0    0.0    0.0    0.0    2.0    1.0   \n",
       "2   MSTRG.100.1  476.0  596.5  487.5  492.0  431.0  409.0  376.0  479.0   \n",
       "3   MSTRG.100.2   33.0    0.0   16.5    0.0   33.0    0.0    0.0   31.0   \n",
       "4  MSTRG.1000.1    5.0    2.0    7.0    5.0    6.0    6.0    3.0    8.0   \n",
       "\n",
       "    bc09   bc10  \n",
       "0    0.0    2.0  \n",
       "1    0.0    0.0  \n",
       "2  326.0  364.0  \n",
       "3    0.0    0.0  \n",
       "4    8.0    9.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_counts_df = pd.read_csv(wd+'noRef_isoQuant_Quantify/noRef_isoquants_quantify_combined_counts_matrix.tsv', sep='\\t')\n",
    "isoquant_counts_df.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(58891, 11)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_counts_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.62.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.62.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.62.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-</td>\n",
       "      <td>u</td>\n",
       "      <td>MSTRG.52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59729</th>\n",
       "      <td>ENSMUSG00000068457.15|ENSMUST00000143958.8</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.29660.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59730</th>\n",
       "      <td>ENSMUSG00000069045.12|ENSMUST00000091190.12</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.29662.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59731</th>\n",
       "      <td>ENSMUSG00000069045.12|ENSMUST00000091190.12</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.29662.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59732</th>\n",
       "      <td>ENSMUSG00000069045.12|ENSMUST00000188182.7</td>\n",
       "      <td>k</td>\n",
       "      <td>MSTRG.29662.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59733</th>\n",
       "      <td>ENSMUSG00000095366.3|ENSMUST00000177893.2</td>\n",
       "      <td>k</td>\n",
       "      <td>MSTRG.29666.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59734 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 reference_gene_id class_code  transcript_id\n",
       "0                                                -          u     MSTRG.49.1\n",
       "1                                                -          u     MSTRG.62.1\n",
       "2                                                -          u     MSTRG.62.2\n",
       "3                                                -          u     MSTRG.62.3\n",
       "4                                                -          u     MSTRG.52.1\n",
       "...                                            ...        ...            ...\n",
       "59729   ENSMUSG00000068457.15|ENSMUST00000143958.8          j  MSTRG.29660.1\n",
       "59730  ENSMUSG00000069045.12|ENSMUST00000091190.12          =  MSTRG.29662.1\n",
       "59731  ENSMUSG00000069045.12|ENSMUST00000091190.12          j  MSTRG.29662.2\n",
       "59732   ENSMUSG00000069045.12|ENSMUST00000188182.7          k  MSTRG.29662.3\n",
       "59733    ENSMUSG00000095366.3|ENSMUST00000177893.2          k  MSTRG.29666.1\n",
       "\n",
       "[59734 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tracking_df[['reference_gene_id','class_code','transcript_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58891, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform_id</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58888</th>\n",
       "      <td>__ambiguous</td>\n",
       "      <td>130040.0</td>\n",
       "      <td>152930.0</td>\n",
       "      <td>127944.0</td>\n",
       "      <td>149126.0</td>\n",
       "      <td>123330.0</td>\n",
       "      <td>110575.0</td>\n",
       "      <td>97357.0</td>\n",
       "      <td>148237.0</td>\n",
       "      <td>113742.0</td>\n",
       "      <td>125599.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58889</th>\n",
       "      <td>__no_feature</td>\n",
       "      <td>92337.0</td>\n",
       "      <td>115778.0</td>\n",
       "      <td>122255.0</td>\n",
       "      <td>120281.0</td>\n",
       "      <td>129410.0</td>\n",
       "      <td>103025.0</td>\n",
       "      <td>108810.0</td>\n",
       "      <td>107864.0</td>\n",
       "      <td>89838.0</td>\n",
       "      <td>110178.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58890</th>\n",
       "      <td>__not_aligned</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          isoform_id      bc01      bc02      bc03      bc04      bc05  \\\n",
       "58888    __ambiguous  130040.0  152930.0  127944.0  149126.0  123330.0   \n",
       "58889   __no_feature   92337.0  115778.0  122255.0  120281.0  129410.0   \n",
       "58890  __not_aligned       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "           bc06      bc07      bc08      bc09      bc10  \n",
       "58888  110575.0   97357.0  148237.0  113742.0  125599.0  \n",
       "58889  103025.0  108810.0  107864.0   89838.0  110178.0  \n",
       "58890       0.0       0.0       0.0       0.0       0.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(isoquant_counts_df.shape)\n",
    "# 59734 \n",
    "isoquant_counts_df[isoquant_counts_df.isoform_id.str.contains('_')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids=isoquant_counts_df[~isoquant_counts_df.isoform_id.str.contains('_')].merge(tracking_df[['reference_gene_id','class_code','transcript_id']], \\\n",
    "                         how='left', left_on='isoform_id', right_on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform_id</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "      <th>reference_gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4895</th>\n",
       "      <td>MSTRG.12284.4</td>\n",
       "      <td>6.25</td>\n",
       "      <td>7.10</td>\n",
       "      <td>4.65</td>\n",
       "      <td>9.2</td>\n",
       "      <td>5.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ENSMUSG00000023927.16|ENSMUST00000169480.8</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.12284.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18925</th>\n",
       "      <td>MSTRG.18675.1</td>\n",
       "      <td>8.67</td>\n",
       "      <td>12.67</td>\n",
       "      <td>7.00</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.67</td>\n",
       "      <td>8.67</td>\n",
       "      <td>9.50</td>\n",
       "      <td>8.67</td>\n",
       "      <td>6.33</td>\n",
       "      <td>12.83</td>\n",
       "      <td>ENSMUSG00000028397.14|ENSMUST00000030102.12</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.18675.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20416</th>\n",
       "      <td>MSTRG.19411.3</td>\n",
       "      <td>45.67</td>\n",
       "      <td>62.33</td>\n",
       "      <td>42.33</td>\n",
       "      <td>57.0</td>\n",
       "      <td>45.33</td>\n",
       "      <td>40.67</td>\n",
       "      <td>41.00</td>\n",
       "      <td>62.00</td>\n",
       "      <td>50.33</td>\n",
       "      <td>66.33</td>\n",
       "      <td>ENSMUSG00000028782.15|ENSMUST00000106017.8</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.19411.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49034</th>\n",
       "      <td>MSTRG.5415.3</td>\n",
       "      <td>53.36</td>\n",
       "      <td>63.29</td>\n",
       "      <td>70.40</td>\n",
       "      <td>59.6</td>\n",
       "      <td>45.40</td>\n",
       "      <td>114.50</td>\n",
       "      <td>88.86</td>\n",
       "      <td>135.02</td>\n",
       "      <td>78.69</td>\n",
       "      <td>121.38</td>\n",
       "      <td>ENSMUSG00000025137.16|ENSMUST00000106188.4</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.5415.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51427</th>\n",
       "      <td>MSTRG.6514.1</td>\n",
       "      <td>28.00</td>\n",
       "      <td>29.50</td>\n",
       "      <td>22.50</td>\n",
       "      <td>14.5</td>\n",
       "      <td>12.50</td>\n",
       "      <td>16.50</td>\n",
       "      <td>15.50</td>\n",
       "      <td>15.50</td>\n",
       "      <td>16.00</td>\n",
       "      <td>22.00</td>\n",
       "      <td>ENSMUSG00000037679.10|ENSMUST00000222275.2</td>\n",
       "      <td>k</td>\n",
       "      <td>MSTRG.6514.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51428</th>\n",
       "      <td>MSTRG.6514.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>13.50</td>\n",
       "      <td>6.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.00</td>\n",
       "      <td>14.00</td>\n",
       "      <td>ENSMUSG00000037679.10|ENSMUST00000101029.4</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.6514.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          isoform_id   bc01   bc02   bc03  bc04   bc05    bc06   bc07    bc08  \\\n",
       "4895   MSTRG.12284.4   6.25   7.10   4.65   9.2   5.80    0.00   0.00    0.00   \n",
       "18925  MSTRG.18675.1   8.67  12.67   7.00  10.0   8.67    8.67   9.50    8.67   \n",
       "20416  MSTRG.19411.3  45.67  62.33  42.33  57.0  45.33   40.67  41.00   62.00   \n",
       "49034   MSTRG.5415.3  53.36  63.29  70.40  59.6  45.40  114.50  88.86  135.02   \n",
       "51427   MSTRG.6514.1  28.00  29.50  22.50  14.5  12.50   16.50  15.50   15.50   \n",
       "51428   MSTRG.6514.2   0.00   0.00   0.00   0.0   0.00   13.50   6.50    0.00   \n",
       "\n",
       "        bc09    bc10                            reference_gene_id class_code  \\\n",
       "4895    0.00    0.00   ENSMUSG00000023927.16|ENSMUST00000169480.8          =   \n",
       "18925   6.33   12.83  ENSMUSG00000028397.14|ENSMUST00000030102.12          =   \n",
       "20416  50.33   66.33   ENSMUSG00000028782.15|ENSMUST00000106017.8          j   \n",
       "49034  78.69  121.38   ENSMUSG00000025137.16|ENSMUST00000106188.4          =   \n",
       "51427  16.00   22.00   ENSMUSG00000037679.10|ENSMUST00000222275.2          k   \n",
       "51428   8.00   14.00   ENSMUSG00000037679.10|ENSMUST00000101029.4          j   \n",
       "\n",
       "       transcript_id  \n",
       "4895   MSTRG.12284.4  \n",
       "18925  MSTRG.18675.1  \n",
       "20416  MSTRG.19411.3  \n",
       "49034   MSTRG.5415.3  \n",
       "51427   MSTRG.6514.1  \n",
       "51428   MSTRG.6514.2  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_wRefids.loc[isoquant_wRefids.isoform_id.isin(['MSTRG.5415.3','MSTRG.12284.4', \n",
    "                                                       'MSTRG.6514.2','MSTRG.6514.1',\n",
    "                                                       'MSTRG.18675.1','MSTRG.19411.3'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58888, 14)\n"
     ]
    }
   ],
   "source": [
    "print(isoquant_wRefids.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(isoquant_wRefids.reference_gene_id.head(10))\n",
    "#print(isoquant_wRefids.reference_gene_id.apply(lambda x: x.split('|')[-1]).head(10))\n",
    "#print(isoquant_wRefids.reference_gene_id.apply(lambda x: x.split('|')[0]).head(10))\n",
    "\n",
    "isoquant_wRefids[['ref_gene_id','ref_transcript_id']]=\\\n",
    "isoquant_wRefids['reference_gene_id'].str.split('|',expand=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56015, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform_id</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "      <th>reference_gene_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10.1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSMUSG00000051951.6|ENSMUST00000070533.5</td>\n",
       "      <td>i</td>\n",
       "      <td>MSTRG.10.1</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>ENSMUST00000070533.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100.1</td>\n",
       "      <td>476.00</td>\n",
       "      <td>596.50</td>\n",
       "      <td>487.5</td>\n",
       "      <td>492.00</td>\n",
       "      <td>431.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>479.0</td>\n",
       "      <td>326.00</td>\n",
       "      <td>364.0</td>\n",
       "      <td>ENSMUSG00000067879.4|ENSMUST00000088666.4</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.100.1</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.100.2</td>\n",
       "      <td>33.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSMUSG00000067879.4|ENSMUST00000088666.4</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.100.2</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.1000.1</td>\n",
       "      <td>5.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.00</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.00</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSMUSG00000037035.6|ENSMUST00000038765.6</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.1000.1</td>\n",
       "      <td>ENSMUSG00000037035.6</td>\n",
       "      <td>ENSMUST00000038765.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MSTRG.10000.1</td>\n",
       "      <td>7.00</td>\n",
       "      <td>5.00</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.00</td>\n",
       "      <td>5.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.00</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSMUSG00000099083.8|ENSMUST00000096143.2</td>\n",
       "      <td>k</td>\n",
       "      <td>MSTRG.10000.1</td>\n",
       "      <td>ENSMUSG00000099083.8</td>\n",
       "      <td>ENSMUST00000096143.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58883</th>\n",
       "      <td>MSTRG.9998.1</td>\n",
       "      <td>65.33</td>\n",
       "      <td>70.33</td>\n",
       "      <td>74.0</td>\n",
       "      <td>61.67</td>\n",
       "      <td>51.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>46.33</td>\n",
       "      <td>68.0</td>\n",
       "      <td>ENSMUSG00000001285.13|ENSMUST00000113682.9</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.9998.1</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000113682.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58884</th>\n",
       "      <td>MSTRG.9998.2</td>\n",
       "      <td>18.33</td>\n",
       "      <td>22.33</td>\n",
       "      <td>13.0</td>\n",
       "      <td>17.67</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>14.33</td>\n",
       "      <td>18.0</td>\n",
       "      <td>ENSMUSG00000001285.13|ENSMUST00000171733.8</td>\n",
       "      <td>k</td>\n",
       "      <td>MSTRG.9998.2</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000171733.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58885</th>\n",
       "      <td>MSTRG.9998.3</td>\n",
       "      <td>1.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSMUSG00000001285.13|ENSMUST00000113682.9</td>\n",
       "      <td>n</td>\n",
       "      <td>MSTRG.9998.3</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000113682.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58886</th>\n",
       "      <td>MSTRG.9999.1</td>\n",
       "      <td>24.00</td>\n",
       "      <td>25.00</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.50</td>\n",
       "      <td>22.5</td>\n",
       "      <td>24.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>35.5</td>\n",
       "      <td>26.50</td>\n",
       "      <td>27.0</td>\n",
       "      <td>ENSMUSG00000036678.9|ENSMUST00000041208.9</td>\n",
       "      <td>=</td>\n",
       "      <td>MSTRG.9999.1</td>\n",
       "      <td>ENSMUSG00000036678.9</td>\n",
       "      <td>ENSMUST00000041208.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58887</th>\n",
       "      <td>MSTRG.9999.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSMUSG00000036678.9|ENSMUST00000041208.9</td>\n",
       "      <td>j</td>\n",
       "      <td>MSTRG.9999.2</td>\n",
       "      <td>ENSMUSG00000036678.9</td>\n",
       "      <td>ENSMUST00000041208.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56015 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          isoform_id    bc01    bc02   bc03    bc04   bc05   bc06   bc07  \\\n",
       "1         MSTRG.10.1    0.00    0.00    1.0    0.00    0.0    0.0    2.0   \n",
       "2        MSTRG.100.1  476.00  596.50  487.5  492.00  431.0  409.0  376.0   \n",
       "3        MSTRG.100.2   33.00    0.00   16.5    0.00   33.0    0.0    0.0   \n",
       "4       MSTRG.1000.1    5.00    2.00    7.0    5.00    6.0    6.0    3.0   \n",
       "5      MSTRG.10000.1    7.00    5.00    5.0    5.00    5.0   12.0    1.0   \n",
       "...              ...     ...     ...    ...     ...    ...    ...    ...   \n",
       "58883   MSTRG.9998.1   65.33   70.33   74.0   61.67   51.0   65.0   50.0   \n",
       "58884   MSTRG.9998.2   18.33   22.33   13.0   17.67   14.0   14.0   16.0   \n",
       "58885   MSTRG.9998.3    1.33    0.00    0.0    0.00    1.0    1.0    0.0   \n",
       "58886   MSTRG.9999.1   24.00   25.00   26.0   23.50   22.5   24.0   22.5   \n",
       "58887   MSTRG.9999.2    0.00    6.00    0.0    0.00    3.5    0.0    3.5   \n",
       "\n",
       "        bc08    bc09   bc10                           reference_gene_id  \\\n",
       "1        1.0    0.00    0.0   ENSMUSG00000051951.6|ENSMUST00000070533.5   \n",
       "2      479.0  326.00  364.0   ENSMUSG00000067879.4|ENSMUST00000088666.4   \n",
       "3       31.0    0.00    0.0   ENSMUSG00000067879.4|ENSMUST00000088666.4   \n",
       "4        8.0    8.00    9.0   ENSMUSG00000037035.6|ENSMUST00000038765.6   \n",
       "5        9.0    8.00    5.0   ENSMUSG00000099083.8|ENSMUST00000096143.2   \n",
       "...      ...     ...    ...                                         ...   \n",
       "58883   77.0   46.33   68.0  ENSMUSG00000001285.13|ENSMUST00000113682.9   \n",
       "58884   23.0   14.33   18.0  ENSMUSG00000001285.13|ENSMUST00000171733.8   \n",
       "58885    0.0    0.00    0.0  ENSMUSG00000001285.13|ENSMUST00000113682.9   \n",
       "58886   35.5   26.50   27.0   ENSMUSG00000036678.9|ENSMUST00000041208.9   \n",
       "58887    0.0    0.00    0.0   ENSMUSG00000036678.9|ENSMUST00000041208.9   \n",
       "\n",
       "      class_code  transcript_id            ref_gene_id     ref_transcript_id  \n",
       "1              i     MSTRG.10.1   ENSMUSG00000051951.6  ENSMUST00000070533.5  \n",
       "2              =    MSTRG.100.1   ENSMUSG00000067879.4  ENSMUST00000088666.4  \n",
       "3              j    MSTRG.100.2   ENSMUSG00000067879.4  ENSMUST00000088666.4  \n",
       "4              =   MSTRG.1000.1   ENSMUSG00000037035.6  ENSMUST00000038765.6  \n",
       "5              k  MSTRG.10000.1   ENSMUSG00000099083.8  ENSMUST00000096143.2  \n",
       "...          ...            ...                    ...                   ...  \n",
       "58883          =   MSTRG.9998.1  ENSMUSG00000001285.13  ENSMUST00000113682.9  \n",
       "58884          k   MSTRG.9998.2  ENSMUSG00000001285.13  ENSMUST00000171733.8  \n",
       "58885          n   MSTRG.9998.3  ENSMUSG00000001285.13  ENSMUST00000113682.9  \n",
       "58886          =   MSTRG.9999.1   ENSMUSG00000036678.9  ENSMUST00000041208.9  \n",
       "58887          j   MSTRG.9999.2   ENSMUSG00000036678.9  ENSMUST00000041208.9  \n",
       "\n",
       "[56015 rows x 16 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(isoquant_wRefids.loc[isoquant_wRefids['reference_gene_id']!='-',].shape)\n",
    "isoquant_wRefids.loc[isoquant_wRefids['reference_gene_id']!='-',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "28",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids=isoquant_wRefids[['transcript_id','ref_gene_id','ref_transcript_id','class_code',\\\n",
    "                  'bc01','bc02','bc03','bc04','bc05',\\\n",
    "                 'bc06','bc07','bc08','bc09','bc10']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cda82ad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids.to_csv(wd+'noRef_isoQuant_Quantify/isoquant_counts_mapped.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58888, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.1.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10.1</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>ENSMUST00000070533.5</td>\n",
       "      <td>i</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100.1</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>=</td>\n",
       "      <td>476.0</td>\n",
       "      <td>596.5</td>\n",
       "      <td>487.5</td>\n",
       "      <td>492.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>479.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>364.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.100.2</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>j</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.1000.1</td>\n",
       "      <td>ENSMUSG00000037035.6</td>\n",
       "      <td>ENSMUST00000038765.6</td>\n",
       "      <td>=</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id           ref_gene_id     ref_transcript_id class_code   bc01  \\\n",
       "0     MSTRG.1.1                     -                  None          u    0.0   \n",
       "1    MSTRG.10.1  ENSMUSG00000051951.6  ENSMUST00000070533.5          i    0.0   \n",
       "2   MSTRG.100.1  ENSMUSG00000067879.4  ENSMUST00000088666.4          =  476.0   \n",
       "3   MSTRG.100.2  ENSMUSG00000067879.4  ENSMUST00000088666.4          j   33.0   \n",
       "4  MSTRG.1000.1  ENSMUSG00000037035.6  ENSMUST00000038765.6          =    5.0   \n",
       "\n",
       "    bc02   bc03   bc04   bc05   bc06   bc07   bc08   bc09   bc10  \n",
       "0    0.0    0.0    1.0    1.0    0.0    0.0    0.0    0.0    2.0  \n",
       "1    0.0    1.0    0.0    0.0    0.0    2.0    1.0    0.0    0.0  \n",
       "2  596.5  487.5  492.0  431.0  409.0  376.0  479.0  326.0  364.0  \n",
       "3    0.0   16.5    0.0   33.0    0.0    0.0   31.0    0.0    0.0  \n",
       "4    2.0    7.0    5.0    6.0    6.0    3.0    8.0    8.0    9.0  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(isoquant_wRefids.shape)\n",
    "isoquant_wRefids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7bc0c9bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.1.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10.1</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>ENSMUST00000070533.5</td>\n",
       "      <td>i</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100.1</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>=</td>\n",
       "      <td>476.00</td>\n",
       "      <td>596.50</td>\n",
       "      <td>487.5</td>\n",
       "      <td>492.00</td>\n",
       "      <td>431.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>479.0</td>\n",
       "      <td>326.00</td>\n",
       "      <td>364.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.100.2</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>j</td>\n",
       "      <td>33.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.1000.1</td>\n",
       "      <td>ENSMUSG00000037035.6</td>\n",
       "      <td>ENSMUST00000038765.6</td>\n",
       "      <td>=</td>\n",
       "      <td>5.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.00</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.00</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58883</th>\n",
       "      <td>MSTRG.9998.1</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000113682.9</td>\n",
       "      <td>=</td>\n",
       "      <td>65.33</td>\n",
       "      <td>70.33</td>\n",
       "      <td>74.0</td>\n",
       "      <td>61.67</td>\n",
       "      <td>51.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>46.33</td>\n",
       "      <td>68.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58884</th>\n",
       "      <td>MSTRG.9998.2</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000171733.8</td>\n",
       "      <td>k</td>\n",
       "      <td>18.33</td>\n",
       "      <td>22.33</td>\n",
       "      <td>13.0</td>\n",
       "      <td>17.67</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>14.33</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58885</th>\n",
       "      <td>MSTRG.9998.3</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000113682.9</td>\n",
       "      <td>n</td>\n",
       "      <td>1.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58886</th>\n",
       "      <td>MSTRG.9999.1</td>\n",
       "      <td>ENSMUSG00000036678.9</td>\n",
       "      <td>ENSMUST00000041208.9</td>\n",
       "      <td>=</td>\n",
       "      <td>24.00</td>\n",
       "      <td>25.00</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.50</td>\n",
       "      <td>22.5</td>\n",
       "      <td>24.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>35.5</td>\n",
       "      <td>26.50</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58887</th>\n",
       "      <td>MSTRG.9999.2</td>\n",
       "      <td>ENSMUSG00000036678.9</td>\n",
       "      <td>ENSMUST00000041208.9</td>\n",
       "      <td>j</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58888 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id            ref_gene_id     ref_transcript_id class_code  \\\n",
       "0         MSTRG.1.1                      -                  None          u   \n",
       "1        MSTRG.10.1   ENSMUSG00000051951.6  ENSMUST00000070533.5          i   \n",
       "2       MSTRG.100.1   ENSMUSG00000067879.4  ENSMUST00000088666.4          =   \n",
       "3       MSTRG.100.2   ENSMUSG00000067879.4  ENSMUST00000088666.4          j   \n",
       "4      MSTRG.1000.1   ENSMUSG00000037035.6  ENSMUST00000038765.6          =   \n",
       "...             ...                    ...                   ...        ...   \n",
       "58883  MSTRG.9998.1  ENSMUSG00000001285.13  ENSMUST00000113682.9          =   \n",
       "58884  MSTRG.9998.2  ENSMUSG00000001285.13  ENSMUST00000171733.8          k   \n",
       "58885  MSTRG.9998.3  ENSMUSG00000001285.13  ENSMUST00000113682.9          n   \n",
       "58886  MSTRG.9999.1   ENSMUSG00000036678.9  ENSMUST00000041208.9          =   \n",
       "58887  MSTRG.9999.2   ENSMUSG00000036678.9  ENSMUST00000041208.9          j   \n",
       "\n",
       "         bc01    bc02   bc03    bc04   bc05   bc06   bc07   bc08    bc09  \\\n",
       "0        0.00    0.00    0.0    1.00    1.0    0.0    0.0    0.0    0.00   \n",
       "1        0.00    0.00    1.0    0.00    0.0    0.0    2.0    1.0    0.00   \n",
       "2      476.00  596.50  487.5  492.00  431.0  409.0  376.0  479.0  326.00   \n",
       "3       33.00    0.00   16.5    0.00   33.0    0.0    0.0   31.0    0.00   \n",
       "4        5.00    2.00    7.0    5.00    6.0    6.0    3.0    8.0    8.00   \n",
       "...       ...     ...    ...     ...    ...    ...    ...    ...     ...   \n",
       "58883   65.33   70.33   74.0   61.67   51.0   65.0   50.0   77.0   46.33   \n",
       "58884   18.33   22.33   13.0   17.67   14.0   14.0   16.0   23.0   14.33   \n",
       "58885    1.33    0.00    0.0    0.00    1.0    1.0    0.0    0.0    0.00   \n",
       "58886   24.00   25.00   26.0   23.50   22.5   24.0   22.5   35.5   26.50   \n",
       "58887    0.00    6.00    0.0    0.00    3.5    0.0    3.5    0.0    0.00   \n",
       "\n",
       "        bc10  \n",
       "0        2.0  \n",
       "1        0.0  \n",
       "2      364.0  \n",
       "3        0.0  \n",
       "4        9.0  \n",
       "...      ...  \n",
       "58883   68.0  \n",
       "58884   18.0  \n",
       "58885    0.0  \n",
       "58886   27.0  \n",
       "58887    0.0  \n",
       "\n",
       "[58888 rows x 14 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_wRefids.transcript_id.split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e1246da",
   "metadata": {},
   "outputs": [],
   "source": [
    "#new_transcript_id= MSTRG.10.1|ENSMUST00000070533.5|i\n",
    "#new_gene_id= MSTRG.10|ENSMUSG00000051951.6\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "fabc5c27",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids['strgt_gene_ids']=\\\n",
    "isoquant_wRefids.transcript_id.apply(lambda x: ('.').join(x.split('.')[0:2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "05d58698",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strgt_gene_ids\n",
       "MSTRG.1        1\n",
       "MSTRG.10       1\n",
       "MSTRG.100      1\n",
       "MSTRG.1000     1\n",
       "MSTRG.10000    1\n",
       "              ..\n",
       "MSTRG.9995     1\n",
       "MSTRG.9996     1\n",
       "MSTRG.9997     1\n",
       "MSTRG.9998     1\n",
       "MSTRG.9999     1\n",
       "Name: ref_gene_id, Length: 29622, dtype: int64"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(isoquant_wRefids[['ref_gene_id','strgt_gene_ids']]\\\n",
    " .groupby('strgt_gene_ids')['ref_gene_id'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "c626867f",
   "metadata": {},
   "outputs": [],
   "source": [
    "test=isoquant_wRefids.groupby('strgt_gene_ids').agg({'ref_gene_id':lambda x: len(pd.unique(x))})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "570fec62",
   "metadata": {},
   "outputs": [],
   "source": [
    "test=test.add_suffix('_Count').reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "f3524d3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strgt_gene_ids</th>\n",
       "      <th>ref_gene_id_Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.1000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.10000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29617</th>\n",
       "      <td>MSTRG.9995</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29618</th>\n",
       "      <td>MSTRG.9996</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29619</th>\n",
       "      <td>MSTRG.9997</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29620</th>\n",
       "      <td>MSTRG.9998</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29621</th>\n",
       "      <td>MSTRG.9999</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29622 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      strgt_gene_ids  ref_gene_id_Count\n",
       "0            MSTRG.1                  1\n",
       "1           MSTRG.10                  1\n",
       "2          MSTRG.100                  1\n",
       "3         MSTRG.1000                  1\n",
       "4        MSTRG.10000                  1\n",
       "...              ...                ...\n",
       "29617     MSTRG.9995                  1\n",
       "29618     MSTRG.9996                  1\n",
       "29619     MSTRG.9997                  1\n",
       "29620     MSTRG.9998                  1\n",
       "29621     MSTRG.9999                  1\n",
       "\n",
       "[29622 rows x 2 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "439d8455",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60       MSTRG.10051\n",
       "76       MSTRG.10066\n",
       "304      MSTRG.10271\n",
       "322      MSTRG.10288\n",
       "363      MSTRG.10324\n",
       "            ...     \n",
       "29166     MSTRG.9588\n",
       "29282     MSTRG.9692\n",
       "29304     MSTRG.9711\n",
       "29359     MSTRG.9761\n",
       "29441     MSTRG.9836\n",
       "Name: strgt_gene_ids, Length: 455, dtype: object"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test[test.ref_gene_id_Count>1].strgt_gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "419b7624",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "455"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isoquant_wRefids.loc\\\n",
    "[isoquant_wRefids.strgt_gene_ids.isin(test[test.ref_gene_id_Count>1].strgt_gene_ids)]\\\n",
    ".strgt_gene_ids.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "9a3192b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strgt_gene_ids</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>MSTRG.10051</td>\n",
       "      <td>MSTRG.10051.1</td>\n",
       "      <td>ENSMUSG00000005982.15</td>\n",
       "      <td>ENSMUST00000186375.8</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>MSTRG.10051</td>\n",
       "      <td>MSTRG.10051.10</td>\n",
       "      <td>ENSMUSG00000005982.15</td>\n",
       "      <td>ENSMUST00000186375.8</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>MSTRG.10051</td>\n",
       "      <td>MSTRG.10051.11</td>\n",
       "      <td>ENSMUSG00000014232.15</td>\n",
       "      <td>ENSMUST00000040881.14</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>MSTRG.10051</td>\n",
       "      <td>MSTRG.10051.12</td>\n",
       "      <td>ENSMUSG00000014232.15</td>\n",
       "      <td>ENSMUST00000040881.14</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>MSTRG.10051</td>\n",
       "      <td>MSTRG.10051.13</td>\n",
       "      <td>ENSMUSG00000014232.15</td>\n",
       "      <td>ENSMUST00000040881.14</td>\n",
       "      <td>=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58494</th>\n",
       "      <td>MSTRG.9836</td>\n",
       "      <td>MSTRG.9836.2</td>\n",
       "      <td>ENSMUSG00000045608.8</td>\n",
       "      <td>ENSMUST00000054244.7</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58495</th>\n",
       "      <td>MSTRG.9836</td>\n",
       "      <td>MSTRG.9836.3</td>\n",
       "      <td>ENSMUSG00000045608.8</td>\n",
       "      <td>ENSMUST00000054244.7</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58496</th>\n",
       "      <td>MSTRG.9836</td>\n",
       "      <td>MSTRG.9836.4</td>\n",
       "      <td>ENSMUSG00000045608.8</td>\n",
       "      <td>ENSMUST00000054244.7</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58497</th>\n",
       "      <td>MSTRG.9836</td>\n",
       "      <td>MSTRG.9836.5</td>\n",
       "      <td>ENSMUSG00000115699.3</td>\n",
       "      <td>ENSMUST00000243426.1</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58498</th>\n",
       "      <td>MSTRG.9836</td>\n",
       "      <td>MSTRG.9836.6</td>\n",
       "      <td>ENSMUSG00000115699.3</td>\n",
       "      <td>ENSMUST00000243426.1</td>\n",
       "      <td>j</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2569 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      strgt_gene_ids   transcript_id            ref_gene_id  \\\n",
       "134      MSTRG.10051   MSTRG.10051.1  ENSMUSG00000005982.15   \n",
       "135      MSTRG.10051  MSTRG.10051.10  ENSMUSG00000005982.15   \n",
       "136      MSTRG.10051  MSTRG.10051.11  ENSMUSG00000014232.15   \n",
       "137      MSTRG.10051  MSTRG.10051.12  ENSMUSG00000014232.15   \n",
       "138      MSTRG.10051  MSTRG.10051.13  ENSMUSG00000014232.15   \n",
       "...              ...             ...                    ...   \n",
       "58494     MSTRG.9836    MSTRG.9836.2   ENSMUSG00000045608.8   \n",
       "58495     MSTRG.9836    MSTRG.9836.3   ENSMUSG00000045608.8   \n",
       "58496     MSTRG.9836    MSTRG.9836.4   ENSMUSG00000045608.8   \n",
       "58497     MSTRG.9836    MSTRG.9836.5   ENSMUSG00000115699.3   \n",
       "58498     MSTRG.9836    MSTRG.9836.6   ENSMUSG00000115699.3   \n",
       "\n",
       "           ref_transcript_id class_code  \n",
       "134     ENSMUST00000186375.8          j  \n",
       "135     ENSMUST00000186375.8          j  \n",
       "136    ENSMUST00000040881.14          j  \n",
       "137    ENSMUST00000040881.14          j  \n",
       "138    ENSMUST00000040881.14          =  \n",
       "...                      ...        ...  \n",
       "58494   ENSMUST00000054244.7          j  \n",
       "58495   ENSMUST00000054244.7          j  \n",
       "58496   ENSMUST00000054244.7          j  \n",
       "58497   ENSMUST00000243426.1          j  \n",
       "58498   ENSMUST00000243426.1          j  \n",
       "\n",
       "[2569 rows x 5 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_wRefids.loc\\\n",
    "[isoquant_wRefids.strgt_gene_ids.isin(test[test.ref_gene_id_Count>1].strgt_gene_ids),\\\n",
    " ['strgt_gene_ids','transcript_id','ref_gene_id','ref_transcript_id','class_code']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58888"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isoquant_wRefids.transcript_id.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31930"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isoquant_wRefids.ref_transcript_id.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "6abe8bd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10680"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isoquant_wRefids.loc[isoquant_wRefids.\\\n",
    "                     duplicated(subset=['ref_transcript_id'])].ref_transcript_id.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16864</th>\n",
       "      <td>MSTRG.17696.2</td>\n",
       "      <td>ENSMUSG00000000001.5</td>\n",
       "      <td>ENSMUST00000000001.5</td>\n",
       "      <td>j</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>8.50</td>\n",
       "      <td>6.50</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.50</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.50</td>\n",
       "      <td>7.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48356</th>\n",
       "      <td>MSTRG.5145.3</td>\n",
       "      <td>ENSMUSG00000000049.12</td>\n",
       "      <td>ENSMUST00000000049.6</td>\n",
       "      <td>j</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48355</th>\n",
       "      <td>MSTRG.5145.2</td>\n",
       "      <td>ENSMUSG00000000049.12</td>\n",
       "      <td>ENSMUST00000000049.6</td>\n",
       "      <td>j</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26078</th>\n",
       "      <td>MSTRG.21976.3</td>\n",
       "      <td>ENSMUSG00000000058.7</td>\n",
       "      <td>ENSMUST00000000058.7</td>\n",
       "      <td>j</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51592</th>\n",
       "      <td>MSTRG.6581.2</td>\n",
       "      <td>ENSMUSG00000000078.8</td>\n",
       "      <td>ENSMUST00000000080.8</td>\n",
       "      <td>m</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.33</td>\n",
       "      <td>9.67</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.33</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.33</td>\n",
       "      <td>4.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58818</th>\n",
       "      <td>MSTRG.9966.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58828</th>\n",
       "      <td>MSTRG.9971.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58842</th>\n",
       "      <td>MSTRG.998.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58846</th>\n",
       "      <td>MSTRG.9981.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58861</th>\n",
       "      <td>MSTRG.9989.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26958 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id            ref_gene_id     ref_transcript_id class_code  \\\n",
       "16864  MSTRG.17696.2   ENSMUSG00000000001.5  ENSMUST00000000001.5          j   \n",
       "48356   MSTRG.5145.3  ENSMUSG00000000049.12  ENSMUST00000000049.6          j   \n",
       "48355   MSTRG.5145.2  ENSMUSG00000000049.12  ENSMUST00000000049.6          j   \n",
       "26078  MSTRG.21976.3   ENSMUSG00000000058.7  ENSMUST00000000058.7          j   \n",
       "51592   MSTRG.6581.2   ENSMUSG00000000078.8  ENSMUST00000000080.8          m   \n",
       "...              ...                    ...                   ...        ...   \n",
       "58818   MSTRG.9966.1                      -                  None          u   \n",
       "58828   MSTRG.9971.1                      -                  None          u   \n",
       "58842    MSTRG.998.1                      -                  None          u   \n",
       "58846   MSTRG.9981.1                      -                  None          u   \n",
       "58861   MSTRG.9989.1                      -                  None          u   \n",
       "\n",
       "       bc01  bc02  bc03  bc04  bc05  bc06  bc07  bc08  bc09  bc10  \n",
       "16864   0.0   5.5  8.50  6.50   5.5   6.0  6.50   9.0  6.50  7.50  \n",
       "48356   1.0   3.0  1.00  2.00   2.0   0.0  1.00   1.0  0.00  3.00  \n",
       "48355   0.0   0.0  2.00  1.00   2.0   1.0  1.00   1.0  0.00  0.00  \n",
       "26078   0.0   0.0  0.00  0.00   0.0   1.0  0.00   0.0  0.00  0.00  \n",
       "51592   6.0   7.0  5.33  9.67   7.0   0.0  2.33   8.0  5.33  4.33  \n",
       "...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  \n",
       "58818   0.0   0.0  0.00  0.00   0.0   0.0  2.00   0.0  0.00  0.00  \n",
       "58828   0.0   0.0  0.00  0.00   0.0   1.0  0.00   3.0  0.00  1.00  \n",
       "58842   0.0   0.0  0.00  0.00   0.0   0.0  0.00   0.0  0.00  2.00  \n",
       "58846   0.0   0.0  0.00  0.00   0.0   0.0  1.00   1.0  2.00  0.00  \n",
       "58861   0.0   0.0  0.00  0.00   0.0   2.0  0.00   0.0  0.00  0.00  \n",
       "\n",
       "[26958 rows x 14 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_wRefids.loc[isoquant_wRefids.\\\n",
    "                     duplicated(subset=['ref_transcript_id'])].sort_values(by=['ref_transcript_id'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c06fe50a",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids['uid']= isoquant_wRefids.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class_code\n",
       "=     1997\n",
       "c      511\n",
       "e      136\n",
       "i     5381\n",
       "j    11898\n",
       "k     1241\n",
       "m     1225\n",
       "n      509\n",
       "o      400\n",
       "p       51\n",
       "s        2\n",
       "u     2872\n",
       "x      724\n",
       "y       11\n",
       "dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_wRefids.loc[isoquant_wRefids.\\\n",
    "                     duplicated(subset=['ref_transcript_id'])].groupby(['class_code']).size()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "      <th>strgt_gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16864</th>\n",
       "      <td>MSTRG.17696.2</td>\n",
       "      <td>ENSMUSG00000000001.5</td>\n",
       "      <td>ENSMUST00000000001.5</td>\n",
       "      <td>j</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>8.50</td>\n",
       "      <td>6.50</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.50</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.50</td>\n",
       "      <td>7.50</td>\n",
       "      <td>MSTRG.17696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48356</th>\n",
       "      <td>MSTRG.5145.3</td>\n",
       "      <td>ENSMUSG00000000049.12</td>\n",
       "      <td>ENSMUST00000000049.6</td>\n",
       "      <td>j</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.00</td>\n",
       "      <td>MSTRG.5145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48355</th>\n",
       "      <td>MSTRG.5145.2</td>\n",
       "      <td>ENSMUSG00000000049.12</td>\n",
       "      <td>ENSMUST00000000049.6</td>\n",
       "      <td>j</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MSTRG.5145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26078</th>\n",
       "      <td>MSTRG.21976.3</td>\n",
       "      <td>ENSMUSG00000000058.7</td>\n",
       "      <td>ENSMUST00000000058.7</td>\n",
       "      <td>j</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MSTRG.21976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51592</th>\n",
       "      <td>MSTRG.6581.2</td>\n",
       "      <td>ENSMUSG00000000078.8</td>\n",
       "      <td>ENSMUST00000000080.8</td>\n",
       "      <td>m</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.33</td>\n",
       "      <td>9.67</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.33</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.33</td>\n",
       "      <td>4.33</td>\n",
       "      <td>MSTRG.6581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58818</th>\n",
       "      <td>MSTRG.9966.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MSTRG.9966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58828</th>\n",
       "      <td>MSTRG.9971.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>MSTRG.9971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58842</th>\n",
       "      <td>MSTRG.998.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>MSTRG.998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58846</th>\n",
       "      <td>MSTRG.9981.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MSTRG.9981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58861</th>\n",
       "      <td>MSTRG.9989.1</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MSTRG.9989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26958 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id            ref_gene_id     ref_transcript_id class_code  \\\n",
       "16864  MSTRG.17696.2   ENSMUSG00000000001.5  ENSMUST00000000001.5          j   \n",
       "48356   MSTRG.5145.3  ENSMUSG00000000049.12  ENSMUST00000000049.6          j   \n",
       "48355   MSTRG.5145.2  ENSMUSG00000000049.12  ENSMUST00000000049.6          j   \n",
       "26078  MSTRG.21976.3   ENSMUSG00000000058.7  ENSMUST00000000058.7          j   \n",
       "51592   MSTRG.6581.2   ENSMUSG00000000078.8  ENSMUST00000000080.8          m   \n",
       "...              ...                    ...                   ...        ...   \n",
       "58818   MSTRG.9966.1                      -                  None          u   \n",
       "58828   MSTRG.9971.1                      -                  None          u   \n",
       "58842    MSTRG.998.1                      -                  None          u   \n",
       "58846   MSTRG.9981.1                      -                  None          u   \n",
       "58861   MSTRG.9989.1                      -                  None          u   \n",
       "\n",
       "       bc01  bc02  bc03  bc04  bc05  bc06  bc07  bc08  bc09  bc10  \\\n",
       "16864   0.0   5.5  8.50  6.50   5.5   6.0  6.50   9.0  6.50  7.50   \n",
       "48356   1.0   3.0  1.00  2.00   2.0   0.0  1.00   1.0  0.00  3.00   \n",
       "48355   0.0   0.0  2.00  1.00   2.0   1.0  1.00   1.0  0.00  0.00   \n",
       "26078   0.0   0.0  0.00  0.00   0.0   1.0  0.00   0.0  0.00  0.00   \n",
       "51592   6.0   7.0  5.33  9.67   7.0   0.0  2.33   8.0  5.33  4.33   \n",
       "...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "58818   0.0   0.0  0.00  0.00   0.0   0.0  2.00   0.0  0.00  0.00   \n",
       "58828   0.0   0.0  0.00  0.00   0.0   1.0  0.00   3.0  0.00  1.00   \n",
       "58842   0.0   0.0  0.00  0.00   0.0   0.0  0.00   0.0  0.00  2.00   \n",
       "58846   0.0   0.0  0.00  0.00   0.0   0.0  1.00   1.0  2.00  0.00   \n",
       "58861   0.0   0.0  0.00  0.00   0.0   2.0  0.00   0.0  0.00  0.00   \n",
       "\n",
       "      strgt_gene_ids  \n",
       "16864    MSTRG.17696  \n",
       "48356     MSTRG.5145  \n",
       "48355     MSTRG.5145  \n",
       "26078    MSTRG.21976  \n",
       "51592     MSTRG.6581  \n",
       "...              ...  \n",
       "58818     MSTRG.9966  \n",
       "58828     MSTRG.9971  \n",
       "58842      MSTRG.998  \n",
       "58846     MSTRG.9981  \n",
       "58861     MSTRG.9989  \n",
       "\n",
       "[26958 rows x 15 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_wRefids.loc[isoquant_wRefids.\\\n",
    "                     duplicated(subset=['ref_transcript_id'])].sort_values(by=['ref_transcript_id']) \n",
    "\n",
    "# u = novel gene and novel isoform\n",
    "# c = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f83a67",
   "metadata": {},
   "outputs": [],
   "source": [
    "#new_transcript_id= MSTRG.10051.1|ENSMUSG00000005982.15|ENSMUST00000186375.8|i\n",
    "#new_gene_id=MSTRG.10051\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "795b1ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids=isoquant_wRefids.fillna(value=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "16815a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids['isoform_id']=isoquant_wRefids[['transcript_id', 'ref_gene_id', 'ref_transcript_id','class_code']]\\\n",
    ".agg('|'.join, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "7af1efe7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "      <th>strgt_gene_ids</th>\n",
       "      <th>isoform_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.1.1</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td>u</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MSTRG.1</td>\n",
       "      <td>MSTRG.1.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10.1</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>ENSMUST00000070533.5</td>\n",
       "      <td>i</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MSTRG.10</td>\n",
       "      <td>MSTRG.10.1|ENSMUSG00000051951.6|ENSMUST0000007...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100.1</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>=</td>\n",
       "      <td>476.00</td>\n",
       "      <td>596.50</td>\n",
       "      <td>487.5</td>\n",
       "      <td>492.00</td>\n",
       "      <td>431.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>479.0</td>\n",
       "      <td>326.00</td>\n",
       "      <td>364.0</td>\n",
       "      <td>MSTRG.100</td>\n",
       "      <td>MSTRG.100.1|ENSMUSG00000067879.4|ENSMUST000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.100.2</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>j</td>\n",
       "      <td>33.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MSTRG.100</td>\n",
       "      <td>MSTRG.100.2|ENSMUSG00000067879.4|ENSMUST000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.1000.1</td>\n",
       "      <td>ENSMUSG00000037035.6</td>\n",
       "      <td>ENSMUST00000038765.6</td>\n",
       "      <td>=</td>\n",
       "      <td>5.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.00</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.00</td>\n",
       "      <td>9.0</td>\n",
       "      <td>MSTRG.1000</td>\n",
       "      <td>MSTRG.1000.1|ENSMUSG00000037035.6|ENSMUST00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58883</th>\n",
       "      <td>MSTRG.9998.1</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000113682.9</td>\n",
       "      <td>=</td>\n",
       "      <td>65.33</td>\n",
       "      <td>70.33</td>\n",
       "      <td>74.0</td>\n",
       "      <td>61.67</td>\n",
       "      <td>51.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>46.33</td>\n",
       "      <td>68.0</td>\n",
       "      <td>MSTRG.9998</td>\n",
       "      <td>MSTRG.9998.1|ENSMUSG00000001285.13|ENSMUST0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58884</th>\n",
       "      <td>MSTRG.9998.2</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000171733.8</td>\n",
       "      <td>k</td>\n",
       "      <td>18.33</td>\n",
       "      <td>22.33</td>\n",
       "      <td>13.0</td>\n",
       "      <td>17.67</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>14.33</td>\n",
       "      <td>18.0</td>\n",
       "      <td>MSTRG.9998</td>\n",
       "      <td>MSTRG.9998.2|ENSMUSG00000001285.13|ENSMUST0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58885</th>\n",
       "      <td>MSTRG.9998.3</td>\n",
       "      <td>ENSMUSG00000001285.13</td>\n",
       "      <td>ENSMUST00000113682.9</td>\n",
       "      <td>n</td>\n",
       "      <td>1.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MSTRG.9998</td>\n",
       "      <td>MSTRG.9998.3|ENSMUSG00000001285.13|ENSMUST0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58886</th>\n",
       "      <td>MSTRG.9999.1</td>\n",
       "      <td>ENSMUSG00000036678.9</td>\n",
       "      <td>ENSMUST00000041208.9</td>\n",
       "      <td>=</td>\n",
       "      <td>24.00</td>\n",
       "      <td>25.00</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.50</td>\n",
       "      <td>22.5</td>\n",
       "      <td>24.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>35.5</td>\n",
       "      <td>26.50</td>\n",
       "      <td>27.0</td>\n",
       "      <td>MSTRG.9999</td>\n",
       "      <td>MSTRG.9999.1|ENSMUSG00000036678.9|ENSMUST00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58887</th>\n",
       "      <td>MSTRG.9999.2</td>\n",
       "      <td>ENSMUSG00000036678.9</td>\n",
       "      <td>ENSMUST00000041208.9</td>\n",
       "      <td>j</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MSTRG.9999</td>\n",
       "      <td>MSTRG.9999.2|ENSMUSG00000036678.9|ENSMUST00000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58888 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id            ref_gene_id     ref_transcript_id class_code  \\\n",
       "0         MSTRG.1.1                      -                                u   \n",
       "1        MSTRG.10.1   ENSMUSG00000051951.6  ENSMUST00000070533.5          i   \n",
       "2       MSTRG.100.1   ENSMUSG00000067879.4  ENSMUST00000088666.4          =   \n",
       "3       MSTRG.100.2   ENSMUSG00000067879.4  ENSMUST00000088666.4          j   \n",
       "4      MSTRG.1000.1   ENSMUSG00000037035.6  ENSMUST00000038765.6          =   \n",
       "...             ...                    ...                   ...        ...   \n",
       "58883  MSTRG.9998.1  ENSMUSG00000001285.13  ENSMUST00000113682.9          =   \n",
       "58884  MSTRG.9998.2  ENSMUSG00000001285.13  ENSMUST00000171733.8          k   \n",
       "58885  MSTRG.9998.3  ENSMUSG00000001285.13  ENSMUST00000113682.9          n   \n",
       "58886  MSTRG.9999.1   ENSMUSG00000036678.9  ENSMUST00000041208.9          =   \n",
       "58887  MSTRG.9999.2   ENSMUSG00000036678.9  ENSMUST00000041208.9          j   \n",
       "\n",
       "         bc01    bc02   bc03    bc04   bc05   bc06   bc07   bc08    bc09  \\\n",
       "0        0.00    0.00    0.0    1.00    1.0    0.0    0.0    0.0    0.00   \n",
       "1        0.00    0.00    1.0    0.00    0.0    0.0    2.0    1.0    0.00   \n",
       "2      476.00  596.50  487.5  492.00  431.0  409.0  376.0  479.0  326.00   \n",
       "3       33.00    0.00   16.5    0.00   33.0    0.0    0.0   31.0    0.00   \n",
       "4        5.00    2.00    7.0    5.00    6.0    6.0    3.0    8.0    8.00   \n",
       "...       ...     ...    ...     ...    ...    ...    ...    ...     ...   \n",
       "58883   65.33   70.33   74.0   61.67   51.0   65.0   50.0   77.0   46.33   \n",
       "58884   18.33   22.33   13.0   17.67   14.0   14.0   16.0   23.0   14.33   \n",
       "58885    1.33    0.00    0.0    0.00    1.0    1.0    0.0    0.0    0.00   \n",
       "58886   24.00   25.00   26.0   23.50   22.5   24.0   22.5   35.5   26.50   \n",
       "58887    0.00    6.00    0.0    0.00    3.5    0.0    3.5    0.0    0.00   \n",
       "\n",
       "        bc10 strgt_gene_ids                                         isoform_id  \n",
       "0        2.0        MSTRG.1                                     MSTRG.1.1|-||u  \n",
       "1        0.0       MSTRG.10  MSTRG.10.1|ENSMUSG00000051951.6|ENSMUST0000007...  \n",
       "2      364.0      MSTRG.100  MSTRG.100.1|ENSMUSG00000067879.4|ENSMUST000000...  \n",
       "3        0.0      MSTRG.100  MSTRG.100.2|ENSMUSG00000067879.4|ENSMUST000000...  \n",
       "4        9.0     MSTRG.1000  MSTRG.1000.1|ENSMUSG00000037035.6|ENSMUST00000...  \n",
       "...      ...            ...                                                ...  \n",
       "58883   68.0     MSTRG.9998  MSTRG.9998.1|ENSMUSG00000001285.13|ENSMUST0000...  \n",
       "58884   18.0     MSTRG.9998  MSTRG.9998.2|ENSMUSG00000001285.13|ENSMUST0000...  \n",
       "58885    0.0     MSTRG.9998  MSTRG.9998.3|ENSMUSG00000001285.13|ENSMUST0000...  \n",
       "58886   27.0     MSTRG.9999  MSTRG.9999.1|ENSMUSG00000036678.9|ENSMUST00000...  \n",
       "58887    0.0     MSTRG.9999  MSTRG.9999.2|ENSMUSG00000036678.9|ENSMUST00000...  \n",
       "\n",
       "[58888 rows x 16 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_wRefids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "35",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_counts_combined_wRefids=isoquant_wRefids[['isoform_id',\\\n",
    "                  'bc01','bc02','bc03','bc04','bc05',\\\n",
    "                 'bc06','bc07','bc08','bc09','bc10']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform_id</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.1.1|-||u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10.1|ENSMUSG00000051951.6|ENSMUST00000070533.5|i</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100.1|ENSMUSG00000067879.4|ENSMUST00000088666.4|=</td>\n",
       "      <td>476.0</td>\n",
       "      <td>596.5</td>\n",
       "      <td>487.5</td>\n",
       "      <td>492.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>479.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>364.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.100.2|ENSMUSG00000067879.4|ENSMUST00000088666.4|j</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.1000.1|ENSMUSG00000037035.6|ENSMUST00000038765.6|=</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 isoform_id   bc01   bc02  \\\n",
       "0                                            MSTRG.1.1|-||u    0.0    0.0   \n",
       "1    MSTRG.10.1|ENSMUSG00000051951.6|ENSMUST00000070533.5|i    0.0    0.0   \n",
       "2   MSTRG.100.1|ENSMUSG00000067879.4|ENSMUST00000088666.4|=  476.0  596.5   \n",
       "3   MSTRG.100.2|ENSMUSG00000067879.4|ENSMUST00000088666.4|j   33.0    0.0   \n",
       "4  MSTRG.1000.1|ENSMUSG00000037035.6|ENSMUST00000038765.6|=    5.0    2.0   \n",
       "\n",
       "    bc03   bc04   bc05   bc06   bc07   bc08   bc09   bc10  \n",
       "0    0.0    1.0    1.0    0.0    0.0    0.0    0.0    2.0  \n",
       "1    1.0    0.0    0.0    0.0    2.0    1.0    0.0    0.0  \n",
       "2  487.5  492.0  431.0  409.0  376.0  479.0  326.0  364.0  \n",
       "3   16.5    0.0   33.0    0.0    0.0   31.0    0.0    0.0  \n",
       "4    7.0    5.0    6.0    6.0    3.0    8.0    8.0    9.0  "
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoquant_counts_combined_wRefids.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "38",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_counts_combined_wRefids.to_csv(\\\n",
    "    '/Users/akhorgad/Desktop/MDL/SAryal/noRef_isoformSwitchR/isoquant_counts_combined_wRefids.tsv', \\\n",
    "     sep='\\t') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {},
   "outputs": [],
   "source": [
    "MSTRG.12284.4 ENSMUSG00000023927.16|ENSMUST00000169480.8 (Transcript: ENSMUST00000169480.8 Satb1-213)\n",
    "MSTRG.5415.3 ENSMUSG00000025137.16|ENSMUST00000106188.4 (ENSMUSG00000025137-Pcyt2)\n",
    "\n",
    "MSTRG.19411.3 ENSMUSG00000028782.15|ENSMUST00000106017.8 (Transcript:Adgrb2-205)\n",
    "\n",
    "MSTRG.6514.1 ENSMUSG00000037679.10|ENSMUST00000222275.2 (Transcript Inf2-203) k\n",
    "MSTRG.6514.2 ENSMUSG00000037679.10|ENSMUST00000101029.4 (Transcript Inf2-201) j\n",
    "\n",
    "MSTRG.18675.1 ENSMUSG00000028397.14|ENSMUST00000030102.12 (geneid | transcript id, Kdm4c-201)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_wRefids.isoform_id\n",
    "isoquant_wRefids.loc[isoquant_wRefids.isoform_id.isin(['MSTRG.1.1','MSTRG.10.1','MSTRG.100.1','MSTRG.29666'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42",
   "metadata": {},
   "source": [
    "<h3>Updating the gtf with Ref Ensemble ids</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "43",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_f=wd+'noRef_isoformSwitchR/stringtie_merged.noRef.gtf'\n",
    "# as PyRanges-object\n",
    "gr_noR = pr.read_gtf(gtf_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "try:\n",
    "    db = gffutils.create_db(gtf_f, dbfn='noRef.db')\n",
    "except:\n",
    "    pass\n",
    "db = gffutils.FeatureDB('noRef.db', keep_order=True)\n",
    "with open('noRef.txt', 'w') as fout:\n",
    "    for line in db.all_features():\n",
    "        line = str(line)\n",
    "        \n",
    "        line = line.split(\";\") #make your parsing changes here\n",
    "        fout.write(str(line) + '\\n')\n",
    "db = gffutils.FeatureDB('noRef.db', keep_order=True)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "for line in db.all_features():\n",
    "    print(line)\n",
    "'''\n",
    "\n",
    "    \n",
    "#gene_id \"MSTRG.1\"; transcript_id \"MSTRG.1.1\";  \"\"  \n",
    "#disregard the gene_id part here, only use MSTRG.1.1 and look up gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tmap_df = pd.read_csv(wd+'noRef_isoformSwitchR/noRef.txt', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "49",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_noR_df=gr_noR.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(474332, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>251317</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>251317</td>\n",
       "      <td>251452</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>354605</td>\n",
       "      <td>354787</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>394862</td>\n",
       "      <td>395039</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>422004</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome     Source     Feature   Start     End  Score Strand Frame  \\\n",
       "0  GL456233.2  StringTie  transcript  251317  422545   1000      +     .   \n",
       "1  GL456233.2  StringTie        exon  251317  251452   1000      +     .   \n",
       "2  GL456233.2  StringTie        exon  354605  354787   1000      +     .   \n",
       "3  GL456233.2  StringTie        exon  394862  395039   1000      +     .   \n",
       "4  GL456233.2  StringTie        exon  422004  422545   1000      +     .   \n",
       "\n",
       "    gene_id transcript_id exon_number  \n",
       "0  MSTRG.49    MSTRG.49.1         NaN  \n",
       "1  MSTRG.49    MSTRG.49.1           1  \n",
       "2  MSTRG.49    MSTRG.49.1           2  \n",
       "3  MSTRG.49    MSTRG.49.1           3  \n",
       "4  MSTRG.49    MSTRG.49.1           4  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(gr_noR_df.shape)\n",
    "gr_noR_df.head()\n",
    "# 59734 number of unique transcripts from gffcompare tracking file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(59734,)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr_noR_df[gr_noR_df.Feature=='transcript'].transcript_id.unique().shape\n",
    "# 59734 transcripts in gtf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "ec3cf61c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MSTRG.10051.1', 'MSTRG.10051.2', 'MSTRG.10051.3', 'MSTRG.10051.4',\n",
       "       'MSTRG.10051.5', 'MSTRG.10051.6', 'MSTRG.10051.7', 'MSTRG.10051.8',\n",
       "       'MSTRG.10051.9', 'MSTRG.10051.10', 'MSTRG.10051.11',\n",
       "       'MSTRG.10051.12', 'MSTRG.10051.13', 'MSTRG.10051.14',\n",
       "       'MSTRG.10051.15', 'MSTRG.10051.16', 'MSTRG.10051.17',\n",
       "       'MSTRG.10051.18'], dtype=object)"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#MSTRG.10051.10\n",
    "gr_noR_df.loc[gr_noR_df['gene_id']=='MSTRG.10051',].transcript_id.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58888, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ref_gene_id</th>\n",
       "      <th>ref_transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>bc01</th>\n",
       "      <th>bc02</th>\n",
       "      <th>bc03</th>\n",
       "      <th>bc04</th>\n",
       "      <th>bc05</th>\n",
       "      <th>bc06</th>\n",
       "      <th>bc07</th>\n",
       "      <th>bc08</th>\n",
       "      <th>bc09</th>\n",
       "      <th>bc10</th>\n",
       "      <th>strgt_gene_ids</th>\n",
       "      <th>isoform_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.1.1</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td>u</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MSTRG.1</td>\n",
       "      <td>MSTRG.1.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.10.1</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>ENSMUST00000070533.5</td>\n",
       "      <td>i</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MSTRG.10</td>\n",
       "      <td>MSTRG.10.1|ENSMUSG00000051951.6|ENSMUST00000070533.5|i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.100.1</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>=</td>\n",
       "      <td>476.0</td>\n",
       "      <td>596.5</td>\n",
       "      <td>487.5</td>\n",
       "      <td>492.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>479.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>364.0</td>\n",
       "      <td>MSTRG.100</td>\n",
       "      <td>MSTRG.100.1|ENSMUSG00000067879.4|ENSMUST00000088666.4|=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.100.2</td>\n",
       "      <td>ENSMUSG00000067879.4</td>\n",
       "      <td>ENSMUST00000088666.4</td>\n",
       "      <td>j</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MSTRG.100</td>\n",
       "      <td>MSTRG.100.2|ENSMUSG00000067879.4|ENSMUST00000088666.4|j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.1000.1</td>\n",
       "      <td>ENSMUSG00000037035.6</td>\n",
       "      <td>ENSMUST00000038765.6</td>\n",
       "      <td>=</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>MSTRG.1000</td>\n",
       "      <td>MSTRG.1000.1|ENSMUSG00000037035.6|ENSMUST00000038765.6|=</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id           ref_gene_id     ref_transcript_id class_code   bc01  \\\n",
       "0     MSTRG.1.1                     -                                u    0.0   \n",
       "1    MSTRG.10.1  ENSMUSG00000051951.6  ENSMUST00000070533.5          i    0.0   \n",
       "2   MSTRG.100.1  ENSMUSG00000067879.4  ENSMUST00000088666.4          =  476.0   \n",
       "3   MSTRG.100.2  ENSMUSG00000067879.4  ENSMUST00000088666.4          j   33.0   \n",
       "4  MSTRG.1000.1  ENSMUSG00000037035.6  ENSMUST00000038765.6          =    5.0   \n",
       "\n",
       "    bc02   bc03   bc04   bc05   bc06   bc07   bc08   bc09   bc10  \\\n",
       "0    0.0    0.0    1.0    1.0    0.0    0.0    0.0    0.0    2.0   \n",
       "1    0.0    1.0    0.0    0.0    0.0    2.0    1.0    0.0    0.0   \n",
       "2  596.5  487.5  492.0  431.0  409.0  376.0  479.0  326.0  364.0   \n",
       "3    0.0   16.5    0.0   33.0    0.0    0.0   31.0    0.0    0.0   \n",
       "4    2.0    7.0    5.0    6.0    6.0    3.0    8.0    8.0    9.0   \n",
       "\n",
       "  strgt_gene_ids                                                isoform_id  \n",
       "0        MSTRG.1                                            MSTRG.1.1|-||u  \n",
       "1       MSTRG.10    MSTRG.10.1|ENSMUSG00000051951.6|ENSMUST00000070533.5|i  \n",
       "2      MSTRG.100   MSTRG.100.1|ENSMUSG00000067879.4|ENSMUST00000088666.4|=  \n",
       "3      MSTRG.100   MSTRG.100.2|ENSMUSG00000067879.4|ENSMUST00000088666.4|j  \n",
       "4     MSTRG.1000  MSTRG.1000.1|ENSMUSG00000037035.6|ENSMUST00000038765.6|=  "
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(isoquant_wRefids.shape)\n",
    "isoquant_wRefids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "1e9fc3bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>isoform_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4804</th>\n",
       "      <td>MSTRG.12246.3</td>\n",
       "      <td>MSTRG.12246.3|ENSMUSG00000023266.12|ENSMUST00000113296.8|=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4892</th>\n",
       "      <td>MSTRG.12284.1</td>\n",
       "      <td>MSTRG.12284.1|ENSMUSG00000023927.16|ENSMUST00000152830.9|=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4893</th>\n",
       "      <td>MSTRG.12284.2</td>\n",
       "      <td>MSTRG.12284.2|ENSMUSG00000023927.16|ENSMUST00000144331.8|=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13480</th>\n",
       "      <td>MSTRG.16165.1</td>\n",
       "      <td>MSTRG.16165.1|ENSMUSG00000073236.6|ENSMUST00000139636.3|=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48040</th>\n",
       "      <td>MSTRG.4988.1</td>\n",
       "      <td>MSTRG.4988.1|ENSMUSG00000020926.17|ENSMUST00000103081.10|=</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id  \\\n",
       "4804   MSTRG.12246.3   \n",
       "4892   MSTRG.12284.1   \n",
       "4893   MSTRG.12284.2   \n",
       "13480  MSTRG.16165.1   \n",
       "48040   MSTRG.4988.1   \n",
       "\n",
       "                                                       isoform_id  \n",
       "4804   MSTRG.12246.3|ENSMUSG00000023266.12|ENSMUST00000113296.8|=  \n",
       "4892   MSTRG.12284.1|ENSMUSG00000023927.16|ENSMUST00000152830.9|=  \n",
       "4893   MSTRG.12284.2|ENSMUSG00000023927.16|ENSMUST00000144331.8|=  \n",
       "13480   MSTRG.16165.1|ENSMUSG00000073236.6|ENSMUST00000139636.3|=  \n",
       "48040  MSTRG.4988.1|ENSMUSG00000020926.17|ENSMUST00000103081.10|=  "
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#MSTRG.12246.3_ENSMUSG00000023266.12_ENSMUST00000113296.8\n",
    "\n",
    "isoquant_wRefids.loc[isoquant_wRefids.isoform_id.str.contains\\\n",
    "                 ('MSTRG.12246.3|MSTRG.12284.1|MSTRG.12284.2|MSTRG.16165.1|MSTRG.4988.1'),['transcript_id','isoform_id']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "82db3564",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "846"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of unique transcripts in the reconstructed gtf : 59734\n",
    "# number of unique transcripts in the quantification matrix based on the gtf : 58888\n",
    "59734 - 58888"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "53",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_wRefids=gr_noR_df.merge(\\\n",
    "    isoquant_wRefids[['transcript_id','strgt_gene_ids','ref_gene_id','ref_transcript_id','isoform_id','class_code']],\\\n",
    "    how='left', on='transcript_id')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(474332, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mstrg_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>mstrg_gene</th>\n",
       "      <th>ensmb_gene</th>\n",
       "      <th>ensmb_transcript</th>\n",
       "      <th>isoform_id</th>\n",
       "      <th>class_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>251317</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "      <td>u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>251317</td>\n",
       "      <td>251452</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "      <td>u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>354605</td>\n",
       "      <td>354787</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>2</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "      <td>u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>394862</td>\n",
       "      <td>395039</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>3</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "      <td>u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>422004</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>MSTRG.49.1</td>\n",
       "      <td>4</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "      <td>u</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome     Source     Feature   Start     End  Score Strand Frame  \\\n",
       "0  GL456233.2  StringTie  transcript  251317  422545   1000      +     .   \n",
       "1  GL456233.2  StringTie        exon  251317  251452   1000      +     .   \n",
       "2  GL456233.2  StringTie        exon  354605  354787   1000      +     .   \n",
       "3  GL456233.2  StringTie        exon  394862  395039   1000      +     .   \n",
       "4  GL456233.2  StringTie        exon  422004  422545   1000      +     .   \n",
       "\n",
       "    gene_id mstrg_transcript exon_number mstrg_gene ensmb_gene  \\\n",
       "0  MSTRG.49       MSTRG.49.1         NaN   MSTRG.49          -   \n",
       "1  MSTRG.49       MSTRG.49.1           1   MSTRG.49          -   \n",
       "2  MSTRG.49       MSTRG.49.1           2   MSTRG.49          -   \n",
       "3  MSTRG.49       MSTRG.49.1           3   MSTRG.49          -   \n",
       "4  MSTRG.49       MSTRG.49.1           4   MSTRG.49          -   \n",
       "\n",
       "  ensmb_transcript       isoform_id class_code  \n",
       "0                   MSTRG.49.1|-||u          u  \n",
       "1                   MSTRG.49.1|-||u          u  \n",
       "2                   MSTRG.49.1|-||u          u  \n",
       "3                   MSTRG.49.1|-||u          u  \n",
       "4                   MSTRG.49.1|-||u          u  "
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(gr_wRefids.shape)\n",
    "gr_wRefids.head(5)\n",
    "# 7618 transcript ids in the gtf do not exist in isoquant results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class_code\n",
       "=    178772\n",
       "c     10856\n",
       "e      1180\n",
       "i     13420\n",
       "j    157268\n",
       "k     70125\n",
       "m     13966\n",
       "n      5578\n",
       "o      2821\n",
       "p       864\n",
       "s        14\n",
       "u      6995\n",
       "x      4718\n",
       "y       137\n",
       "dtype: int64"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr_wRefids[['ref_transcript_id','class_code']].groupby(['class_code']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class_code\n",
       "u    6995\n",
       "dtype: int64"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr_wRefids[gr_wRefids['ref_transcript_id']==\"\"].groupby(['class_code']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "6bda6e70",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_wRefids=gr_wRefids.rename(columns={'transcript_id':'mstrg_transcript'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "58",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_wRefids=gr_wRefids.rename(columns={'strgt_gene_ids':'mstrg_gene',\n",
    "                  'isoform_id':'transcript_id',                    \n",
    "                  'ref_gene_id': 'ensmb_gene',\n",
    "                  'ref_transcript_id':'ensmb_transcript'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7618"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gr_wRefids[gr_wRefids['isoform_id'].isnull()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "c90f62f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_wRefids['isoform_id']=gr_wRefids['isoform_id'].fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "72295c16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gr_wRefids[gr_wRefids['isoform_id'].isnull()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "72c3dc97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7618"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gr_wRefids[gr_wRefids['isoform_id']==\"\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0            MSTRG.49\n",
       "1            MSTRG.49\n",
       "2            MSTRG.49\n",
       "3            MSTRG.49\n",
       "4            MSTRG.49\n",
       "             ...     \n",
       "474327    MSTRG.29666\n",
       "474328    MSTRG.29666\n",
       "474329    MSTRG.29666\n",
       "474330    MSTRG.29666\n",
       "474331    MSTRG.29666\n",
       "Length: 474332, dtype: object"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr_wRefids.apply(lambda x : x.mstrg_transcript if x.isoform_id==\"\" \\\n",
    "                            else x.isoform_id , \\\n",
    "                            axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "61",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_wRefids[\"transcript_id\"]=gr_wRefids.apply(lambda x : x.mstrg_transcript if x.isoform_id==\"\" \\\n",
    "                            else  x.isoform_id, \\\n",
    "                            axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "6ac3be42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gr_wRefids[gr_wRefids['transcript_id']==\"\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2873"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gr_wRefids.loc[(gr_wRefids['class_code']=='u') & (gr_wRefids['Feature']=='transcript'),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "63",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_wRefids=gr_wRefids.drop(['mstrg_gene','mstrg_transcript','ensmb_gene',\\\n",
    "                            'ensmb_transcript','isoform_id','class_code'], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>251317</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>251317</td>\n",
       "      <td>251452</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>1</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>354605</td>\n",
       "      <td>354787</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>2</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>394862</td>\n",
       "      <td>395039</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>3</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>422004</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>4</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474327</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90763694</td>\n",
       "      <td>90772729</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.29666</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474328</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>90763694</td>\n",
       "      <td>90764192</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.29666</td>\n",
       "      <td>1</td>\n",
       "      <td>MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474329</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>90765689</td>\n",
       "      <td>90766299</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.29666</td>\n",
       "      <td>2</td>\n",
       "      <td>MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474330</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>90767289</td>\n",
       "      <td>90767449</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.29666</td>\n",
       "      <td>3</td>\n",
       "      <td>MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474331</th>\n",
       "      <td>chrY</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>90772528</td>\n",
       "      <td>90772729</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.29666</td>\n",
       "      <td>4</td>\n",
       "      <td>MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>474332 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Source     Feature     Start       End  Score Strand  \\\n",
       "0       GL456233.2  StringTie  transcript    251317    422545   1000      +   \n",
       "1       GL456233.2  StringTie        exon    251317    251452   1000      +   \n",
       "2       GL456233.2  StringTie        exon    354605    354787   1000      +   \n",
       "3       GL456233.2  StringTie        exon    394862    395039   1000      +   \n",
       "4       GL456233.2  StringTie        exon    422004    422545   1000      +   \n",
       "...            ...        ...         ...       ...       ...    ...    ...   \n",
       "474327        chrY  StringTie  transcript  90763694  90772729   1000      -   \n",
       "474328        chrY  StringTie        exon  90763694  90764192   1000      -   \n",
       "474329        chrY  StringTie        exon  90765689  90766299   1000      -   \n",
       "474330        chrY  StringTie        exon  90767289  90767449   1000      -   \n",
       "474331        chrY  StringTie        exon  90772528  90772729   1000      -   \n",
       "\n",
       "       Frame      gene_id exon_number  \\\n",
       "0          .     MSTRG.49         NaN   \n",
       "1          .     MSTRG.49           1   \n",
       "2          .     MSTRG.49           2   \n",
       "3          .     MSTRG.49           3   \n",
       "4          .     MSTRG.49           4   \n",
       "...      ...          ...         ...   \n",
       "474327     .  MSTRG.29666         NaN   \n",
       "474328     .  MSTRG.29666           1   \n",
       "474329     .  MSTRG.29666           2   \n",
       "474330     .  MSTRG.29666           3   \n",
       "474331     .  MSTRG.29666           4   \n",
       "\n",
       "                                                    transcript_id  \n",
       "0                                                 MSTRG.49.1|-||u  \n",
       "1                                                 MSTRG.49.1|-||u  \n",
       "2                                                 MSTRG.49.1|-||u  \n",
       "3                                                 MSTRG.49.1|-||u  \n",
       "4                                                 MSTRG.49.1|-||u  \n",
       "...                                                           ...  \n",
       "474327  MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k  \n",
       "474328  MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k  \n",
       "474329  MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k  \n",
       "474330  MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k  \n",
       "474331  MSTRG.29666.1|ENSMUSG00000095366.3|ENSMUST00000177893.2|k  \n",
       "\n",
       "[474332 rows x 11 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr_wRefids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65",
   "metadata": {},
   "source": [
    "# Writing gtf - strtingtie ids "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "66",
   "metadata": {},
   "outputs": [],
   "source": [
    " gr_wRefids_pr= pr.PyRanges(gr_wRefids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>251317</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>251317</td>\n",
       "      <td>251452</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>1</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>354605</td>\n",
       "      <td>354787</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>2</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>394862</td>\n",
       "      <td>395039</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>3</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>422004</td>\n",
       "      <td>422545</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.49</td>\n",
       "      <td>4</td>\n",
       "      <td>MSTRG.49.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>196593</td>\n",
       "      <td>206895</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSTRG.52.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>196593</td>\n",
       "      <td>200564</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.52</td>\n",
       "      <td>1</td>\n",
       "      <td>MSTRG.52.1|-||u</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GL456233.2</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>201259</td>\n",
       "      <td>201382</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MSTRG.52</td>\n",
       "      <td>2</td>\n",
       "      <td>MSTRG.52.1|-||u</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +6    |\n",
       "| (category)   | (object)   | (category)   |   (int64) |   (int64) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| GL456233.2   | StringTie  | transcript   |    251317 |    422545 | ...   |\n",
       "| GL456233.2   | StringTie  | exon         |    251317 |    251452 | ...   |\n",
       "| GL456233.2   | StringTie  | exon         |    354605 |    354787 | ...   |\n",
       "| GL456233.2   | StringTie  | exon         |    394862 |    395039 | ...   |\n",
       "| GL456233.2   | StringTie  | exon         |    422004 |    422545 | ...   |\n",
       "| GL456233.2   | StringTie  | transcript   |    196593 |    206895 | ...   |\n",
       "| GL456233.2   | StringTie  | exon         |    196593 |    200564 | ...   |\n",
       "| GL456233.2   | StringTie  | exon         |    201259 |    201382 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 11 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "6 hidden columns: Score, Strand, Frame, gene_id, exon_number, transcript_id"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr_wRefids_pr.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "68",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_wRefids_pr.to_gtf(\\\n",
    "  \"localdir/stringtie_merged.wRefIds.gtf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GL456233.2\tStringTie\ttranscript\t251318\t422545\t1000\t+\t.\tgene_id \"MSTRG.49\"; transcript_id \"MSTRG.49.1|-||u\";\r\n",
      "GL456233.2\tStringTie\texon\t251318\t251452\t1000\t+\t.\tgene_id \"MSTRG.49\"; exon_number \"1\"; transcript_id \"MSTRG.49.1|-||u\";\r\n",
      "GL456233.2\tStringTie\texon\t354606\t354787\t1000\t+\t.\tgene_id \"MSTRG.49\"; exon_number \"2\"; transcript_id \"MSTRG.49.1|-||u\";\r\n",
      "GL456233.2\tStringTie\texon\t394863\t395039\t1000\t+\t.\tgene_id \"MSTRG.49\"; exon_number \"3\"; transcript_id \"MSTRG.49.1|-||u\";\r\n",
      "GL456233.2\tStringTie\texon\t422005\t422545\t1000\t+\t.\tgene_id \"MSTRG.49\"; exon_number \"4\"; transcript_id \"MSTRG.49.1|-||u\";\r\n",
      "cat: stdout: Broken pipe\r\n"
     ]
    }
   ],
   "source": [
    "! cat /Users/akhorgad/Desktop/MDL/SAryal/noRef_isoformSwitchR/stringtie_merged.noRef.wRefIds.gtf | head -5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "857feaeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "stringtieMergeId^refTranscriptId"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tisoform_id\tbc01\tbc02\tbc03\tbc04\tbc05\tbc06\tbc07\tbc08\tbc09\tbc10\r\n",
      "0\tMSTRG.1.1|-||u\t0.0\t0.0\t0.0\t1.0\t1.0\t0.0\t0.0\t0.0\t0.0\t2.0\r\n",
      "1\tMSTRG.10.1|ENSMUSG00000051951.6|ENSMUST00000070533.5|i\t0.0\t0.0\t1.0\t0.0\t0.0\t0.0\t2.0\t1.0\t0.0\t0.0\r\n",
      "2\tMSTRG.100.1|ENSMUSG00000067879.4|ENSMUST00000088666.4|=\t476.0\t596.5\t487.5\t492.0\t431.0\t409.0\t376.0\t479.0\t326.0\t364.0\r\n",
      "3\tMSTRG.100.2|ENSMUSG00000067879.4|ENSMUST00000088666.4|j\t33.0\t0.0\t16.5\t0.0\t33.0\t0.0\t0.0\t31.0\t0.0\t0.0\r\n",
      "cat: stdout: Broken pipe\r\n"
     ]
    }
   ],
   "source": [
    "! cat data/isoquant_counts_combined_wRefids.tsv | head -5"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
