---
title: "Kinnex Single Cell Vignette"
output: 
---

Installing basic dependencies:
```{r}
install_if_missing <- function(packages) {
    if (length(setdiff(packages, rownames(installed.packages()))) > 0) {
        install.packages(setdiff(packages, rownames(installed.packages())))
    }
}
```

```{r}
install_if_missing(c('tidyverse','stringr','dplyr', 'edgeR', 'ggrepel','DESeq2','Seurat','clustermole'))
```

```{r}
library(tidyverse)
library(DESeq2)
library(dplyr)
```

Analysing sparse matrices created above
```{r}
knitr::opts_chunk$set(echo = TRUE)
    library(tidyverse)
    library(Seurat)
```
PW dir= '/Users/akhorgad/Desktop/MDL/RTD/scIsoquantMatrixBuilder_Results'

Input counts matrix created above from step1:
```{r}
data_dir = "/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/scIsoquantMatrixBuilder_Results/PBMC_BioIVT_10x3p_500cells.genes-sc_matrix_from_isoquant/"

output_prefix = "scKinnex.genes"
```
Reading data in using Read10x() :
```{r}
data = Read10X(data.dir=data_dir,
           gene.column = 1,
           cell.column = 2,
           unique.features = TRUE,
           strip.suffix = FALSE)
```
UMI counts per cell:
```{r}
umi_counts_per_cell = colSums(data)
```
sorting:
```{r}
umi_counts_per_cell = sort(umi_counts_per_cell, decreasing = T)
```
plotting :
```{r}
plot(umi_counts_per_cell, log='xy')
ggsave(filename='umi_counts_per_cell.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/scIsoquantMatrixBuilder_Results', plot = last_plot())
```


Creating seurat object from counts matrix:
```{r}
seurat_obj <- CreateSeuratObject(counts = data, project = "scKinnex", min.cells = 3, min.features = 200)
seurat_obj
```
Terminal Out:
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')An object of class Seurat 
11390 features across 500 samples within 1 assay 
Active assay: RNA (11390 features, 0 variable features)
 1 layer present: counts
 
```{r}
# before filtering
seurat_obj@meta.data %>% summarize(median(nCount_RNA), median(nFeature_RNA))
```
Terminal Out:

median(nCount_RNA)        median(nFeature_RNA)
<dbl>                     <dbl>              
2794.17                  	799	

PercentageFeatureSet - Calculate the percentage of all counts that belong to a given set of features
```{r}
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
```
exploring seurat object:
```{r}
seurat_obj
seurat_obj@meta.data %>% head()

```

UMI counts per cell:
```{r}
seurat_obj@meta.data %>% dplyr::select(nCount_RNA) %>% arrange(desc(nCount_RNA)) %>% mutate(i=row_number()) %>%
  ggplot(aes(x=i, y=nCount_RNA)) + geom_point() + theme_bw() + 
  scale_y_continuous(trans='log10') +
  scale_x_continuous(trans='log10') +
  
  ggtitle("nCount_RNA: UMI counts per cell")

ggsave(filename='nCount_RNA-umi_counts_per_cell.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/scIsoquantMatrixBuilder_Results', plot = last_plot())
```
Feature counts per cell:
```{r}

seurat_obj@meta.data %>% dplyr::select(nFeature_RNA) %>% arrange(desc(nFeature_RNA)) %>% mutate(i=row_number()) %>%
  ggplot(aes(x=i, y=nFeature_RNA)) + geom_point() + theme_bw() + 
  scale_y_continuous(trans='log10') +
  scale_x_continuous(trans='log10') +
  ggtitle("nFeature_RNA: gene count per cell") 

#ggsave(filename='nFeature_RNA-gene_count_per_cell.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/scIsoquantMatrixBuilder_Results', plot = last_plot())

```
# Visualize QC metrics as a violin plot
VlnPlot: Draws a violin plot of single cell data (gene expression, metrics, PC scores, etc.)
```{r}

VlnPlot(seurat_obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
ggsave(filename='VlnPlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/scIsoquantMatrixBuilder_Results', plot = last_plot())

```


```{r}
plot1 <- FeatureScatter(seurat_obj, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(seurat_obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
ggsave(filename='FeatureScatter.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/scIsoquantMatrixBuilder_Results', plot = last_plot())
```
Saving object.RDS

```{r}
# save before filtering

#saveRDS(seurat_obj, file = paste0(data_dir, output_prefix, "-seurat_obj-preCellFiltering.rds"))
```


filtering cells on percent Mitochondria
```{r}
#### filtering cells on 
length(seurat_obj$percent.mt < 15)
seurat_obj <- subset(seurat_obj, 
                     percent.mt < 15)

seurat_obj
```
An object of class Seurat 
11390 features across 447 samples within 1 assay 
Active assay: RNA (11390 features, 0 variable features)
 1 layer present: counts
 
 
Summarize:
```{r}
seurat_obj@meta.data %>% summarize(median(nCount_RNA), median(nFeature_RNA))

```
Temrinal Output:
median(nCount_RNA)      median(nFeature_RNA)
<dbl>                   <int>
2870.08	                817


NormalizeData : Normalize the count data present in a given assay.
Normalization methods =
“librarLogNormalize”: Feature counts for each cell are divided by the total counts for that cell and multiplied by the scale.factor. This is then natural-log transformed using log1p
```{r}
seurat_obj <- NormalizeData(seurat_obj, normalization.method = "LogNormalize", scale.factor = 10000, verbose = TRUE)

```


FindVariableFeatures: Identifies features that are outliers on a 'mean variability plot'.
selection.method =
“vst”: First, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression (loess). Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). Feature variance is then calculated on the standardized values after clipping to a maximum (see clip.max parameter).
```{r}
seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(seurat_obj), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(seurat_obj)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
ggsave(filename='VariableFeatures.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/scIsoquantMatrixBuilder_Results', plot = last_plot())
```


ScaleData: 
Scales and centers features in the dataset. If variables are provided in vars.to.regress, they are individually regressed against each feature, and the resulting residuals are then scaled and centered.
```{r}
all.features <- rownames(seurat_obj)
seurat_obj <- ScaleData(seurat_obj, features = all.features)
```

Performing PCA :

RunPCA: Run Principal Component Analysis on gene expression using IRLBA. For details about stored PCA calculation parameters, see `PrintPCAParams`.

```{r}
seurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(object = seurat_obj))
```

VizDimLoadings: Visualize top genes associated with reduction components:
```{r}

VizDimLoadings(seurat_obj, dims = 1:2, reduction = "pca")
```

DimPlot:
Graphs the output of a dimensional reduction technique (PCA by default). Cells are colored by their identity class.
```{r}

DimPlot(seurat_obj, reduction = "pca") + NoLegend()
#DimHeatmap(seurat_obj, dims = 1, cells = NULL, balanced = TRUE)

ggsave(filename='DimPlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())


# DimHeatmap(
#   object,
#   dims = 1,
#   nfeatures = 30,
#   cells = NULL,
#   reduction = "pca",
#   disp.min = -2.5,
#   disp.max = NULL,
#   balanced = TRUE,
#   projected = FALSE,
#   ncol = NULL,
#   fast = TRUE,
#   raster = TRUE,
#   slot = "scale.data",
#   assays = NULL,
#   combine = TRUE
# )


ElbowPlot(seurat_obj)
ggsave(filename='ElbowPlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())
```
Generating UMAP :
```{r}
seurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)
seurat_obj <- FindClusters(seurat_obj, resolution = 0.5)

seurat_obj <- RunUMAP(seurat_obj, dims = 1:10)
DimPlot(seurat_obj, reduction = "umap")

ggsave(filename='UMAP_DimPlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())

FeaturePlot(seurat_obj, features = c("nFeature_RNA"))
ggsave(filename='nFeature_RNA_FeaturePlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())

FeaturePlot(seurat_obj, features = c("nCount_RNA"))
ggsave(filename='nCount_RNA_FeaturePlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())

FeaturePlot(seurat_obj, features = c("percent.mt"))
ggsave(filename='percent_mt_FeaturePlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())

```
Save Seurat object:
```{r}

saveRDS(seurat_obj, file = paste0(data_dir,output_prefix, "-seurat_obj.rds"))

```

# counts and fractions of cells
```{r}
cluster_counts_n_fracs = seurat_obj@meta.data %>% group_by(seurat_clusters) %>% tally() %>%  mutate(frac=prop.table(n))

cluster_counts_n_fracs

#saveRDS(seurat_obj, file = paste0(output_prefix, "-seurat_obj.rds"))
```

DE, find markers:
 - find markers for every cluster compared to all remaining cells, report only the positive ones
```{r}
#{r}

seurat_obj.markers <- FindAllMarkers(seurat_obj, only.pos = TRUE)
seurat_obj.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

```

```{r}
top_20_markers = seurat_obj.markers %>%
group_by(cluster) %>%
dplyr::filter(avg_log2FC > 1) %>% slice_head(n=20) %>% ungroup()


top_20_markers
```


```{r}
max_cluster <- max(as.numeric(top_20_markers$cluster)) - 1

for (clnum in 0:max_cluster) {
    cluster = top_20_markers %>% filter(cluster == clnum)


        gene.symbols = sapply(cluster$gene, function(x) { str_split(x, "\\^")[[1]][1] })

        gene.symbols = grep("ENSG|ENST|novel", gene.symbols, value=T, invert=T)

    cat(paste0(clnum,":"))
    cat(gene.symbols, sep=",")
    cat("\n")
}
```

Run above list through: http://xteam.xbio.top/ACT to get cell type predictions.


Installing clustermole:
https://cran.rstudio.com/web/packages/clustermole/vignettes/clustermole-intro.html

```{r}
BiocManager::install("igordot/clustermole", update = FALSE)
```

using clustermole to add annotations:

```{r}
library(clustermole)

clustermole_results = NULL

for (clnum in 0:max_cluster) {
  cluster = top_20_markers %>% filter(cluster == clnum)
  
  gene.symbols = sapply(cluster$gene, function(x) { str_split(x, "\\^")[[1]][1] })
  
  gene.symbols = grep("ENSG|ENST|novel", gene.symbols, value=T, invert=T)
  
  
  tryCatch(
    expr = {
      
      cat(paste0(clnum,":"))
    cat(gene.symbols, sep=",")
    cat("\n")
  
    my_overlaps <- clustermole_overlaps(genes = gene.symbols, species = "hs")
  
    clustermole_results = bind_rows(clustermole_results,
                                  my_overlaps %>% mutate(clnum = clnum))
    },
    error = function(e){ 
       message("Error: ", e)
    },
    warning = function(w){
      message("Warning: ", w)
    }
  )
}

```

```{r}

clustermole_summary = clustermole_results  %>% filter(db == "PanglaoDB") %>%
  group_by(clnum) %>% arrange(p_value) %>% filter(row_number() == 1) %>% arrange(clnum) %>%
  ungroup() %>%
  dplyr::select(clnum, organ, celltype, fdr)

clustermole_summary

```

```{r}
write.table(clustermole_summary, file = paste0(data_dir, output_prefix, "-clustermole.tsv"), quote=F, row.names=F, sep="\t")

```


```{r}
# save files for later read/cell tracking

write.table( Idents(seurat_obj), paste0(data_dir, output_prefix, "-cell_cluster_assignments.tsv"), quote=F, row.names=T, sep="\t")


```


```{r}

saveRDS(seurat_obj, file = paste0(data_dir, output_prefix, "-seurat_obj.rds"))

```



Select Tissue type blood
```{r}
# example definition of marker genes for certain cell types

marker_genes = list()

marker_genes[["CD8"]] = c("CD8A","CD8B","GZMB","TRB","TRA","PRF1", "GZMB")

```

# function to extract gene ids with the relevant gene symbols

```{r}
#{r}

# function to extract gene ids with the relevant gene symbols


feature_names = rownames(seurat_obj@assays$RNA$counts)

get_feature_names_with_gene_symbols = function(gene_symbols) {

gene_ids = c()

for (gene_symbol in gene_symbols) {
    found_genes = grep(paste0(gene_symbol,"\\^"), feature_names, value=T)
 if (length(found_genes) > 0) {
        gene_ids = c(gene_ids, found_genes)
    }
}
return(gene_ids)
}

```

# paint umaps according to the features of interest

```{r}
# paint umaps according to the features of interest

feature_ids = get_feature_names_with_gene_symbols(marker_genes[["CD8"]])

VlnPlot(seurat_obj, features = feature_ids)
ggsave(filename='foi_VlnPlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())

FeaturePlot( seurat_obj, features = feature_ids)
ggsave(filename='foi_FeaturePlot.png',path='/Users/akhorgad/Desktop/MDL/RTD/kinnex_sc/', plot = last_plot())


```

